```{r}
source("./all_artifacts.r")
```

### Taxonomic classifications
```{r}
blast <- lapply(
    get_artifact_data("./results/3-Classified", id_key, "BLAST_All"),
    parse_taxonomy
)
sklearn <- lapply(
    get_artifact_data("./results/3-Classified", id_key, "Sklearn"),
    parse_taxonomy
)
```
Simply by counting the number of NA entries in the taxonomy columns, the sklearn
classifier consistently outperforms the traditional blast approach. Furthermore, the sklearn is faster and consumes less memory (aside from the initial training step, which I could not do)

The `known_taxon` function collapses the taxonomy table into the first identification. A similar function will be used to collapse it into at least genus identifications for use with farprotax.

### Phylogenetic trees
```{r}
fasttree <- get_artifact_data(
    "./results/6-RootedTrees", id_key,
    "FastTree_RootedTree"
)
iqtree <- get_artifact_data(
    "./results/6-RootedTrees", id_key,
    "IQTREE_RootedTree"
)
```



### Import frequency tables
The frequency tables are stored in BIOM format: OTUs x sites
```{r}
otu_freqs <- lapply(
    get_artifact_data("./results/2-OTUs", id_key, "otuFreqs"),
    as.data.frame
)
```

The following snippet maps otu ids the genus-level identifications for use with the farprotax database, writing csv files for each.
```{r}
otu_genus <- list()
for (id in names(id_key)) {
    otu_genus[[id]] <- to_genus_csv(otu_freqs[[id]], blast[[id]])
}
combined <- bind_rows(otu_genus) %>%
    arrange(taxon) %>%
    group_by(taxon) %>%
    summarise(across(everything(), sum))
combined[is.na(combined)] <- 0.0
write.csv(combined, "genus_otu_tables.csv", row.names = FALSE)
```

## Diversity analyses

### Alpha diversity


### Beta diversity
```{r}
beta <- get_artifact_data("./results/7-Diversity",
    list(Merged = NULL),
    extension = "",
    metric_list = beta_metrics
)
pcoa2D <- get_artifact_data("./results/8-Analysis",
    list(Merged = NULL),
    extension = "PCOA-2D_",
    metric_list = beta_metrics
)
pcoa2D_merged <- lapply(pcoa2D$Merged, metadata_merge_pcoa, metadata = metadata)
```

After visualizing the beta diversity, we can immediately rule 

PERMANOVA
```{r}
adonis2(beta$Merged$bc ~ metadata$Location)
```

### Functional analysis
The raw output tables here are the absolute abundances of molecular functions in each site using the KEGG Orthology database (KO). PICRUSt2 integrates with other databases, such as Enzyme Commission numbers (EC), but I've chosen to use the KO database because the PICRUSt2 authors describe how they've improved on the database integration compared to PICRUSt1. 
As PICRUSt2 failed to analyze the merged dataset, I will combine the output tables from each sample.
```{r}
ko_all <- ko %>%
    # Merge the PICRUSt2 tables
    reduce(merge, by = "function", all = TRUE) %>%
    as_tibble() %>%
    replace(is.na(.), 0) %>%
    rel_abund() %>% # Convert to relative abundances
    as_tibble() %>%
    sites_x_func() # Transpose into sites x function format
# Compute bray curtis distance
bc_func <- vegdist(ko_all, method = "bray")
pcoa_bc_func <- bc_func %>%
    wcmdscale(k = 2) %>%
    metadata_merge_pcoa(metadata, ., functions = TRUE)
# Compute jaccard distance
ja_func <- vegdist(ko_all, method = "jaccard")
pcoa_ja_func <- ja_func %>%
    wcmdscale(k = 2) %>%
    metadata_merge_pcoa(metadata, ., functions = TRUE)
```

