{
  "cells": [
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "###### Imports\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 54,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Imports\n",
        "from qiime2 import Metadata\n",
        "import pandas as pd\n",
        "from qiime2 import Artifact, sdk\n",
        "from qiime2.plugins.dada2.methods import denoise_pyro # The samples were obtained through pyrosequencing\n",
        "import qiime2.plugins.vsearch.methods as vsm\n",
        "import qiime2.plugins.phylogeny.methods as pm\n",
        "from qiime2.plugins import phylogeny\n",
        "from qiime2.plugins.alignment.methods import mafft\n",
        "import qiime2.plugins.metadata.actions as metadata_actions\n",
        "import qiime2.plugins.feature_classifier.actions as feature_classifier_actions\n",
        "from qiime2.plugins.feature_classifier.pipelines import classify_consensus_vsearch\n",
        "from qiime2.plugins.feature_classifier.pipelines import classify_consensus_blast\n",
        "from qiime2.plugins.feature_table.visualizers import tabulate_seqs\n",
        "import qiime2.plugins.feature_table.actions as fta\n",
        "import matplotlib.pyplot as plt\n",
        "import matplotlib \n",
        "import seaborn as sns\n",
        "from Bio import SeqIO\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 27,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Useful functions\n",
        "matplotlib.use('module://ipykernel.pylab.backend_inline')\n",
        "pm = sdk.PluginManager()\n",
        "def see(artifact):\n",
        "    from_format = artifact.format\n",
        "    if issubclass(from_format, sdk.plugin_manager.SingleFileDirectoryFormatBase):\n",
        "        from_format = artifact.format.file.format\n",
        "    return set(pm.transformers[from_format].keys())\n",
        "import os\n",
        "import pandas\n",
        "import qiime2\n",
        "import tempfile\n",
        "\n",
        "def v2frame(viz_fp: str) -> list:\n",
        "    '''viz_fp is a path to the qiime2 visualization object'''\n",
        "    viz = qiime2.Visualization.load(viz_fp)\n",
        "    with tempfile.TemporaryDirectory() as tmpdir:\n",
        "        viz.export_data(tmpdir)\n",
        "        fp = os.path.join(tmpdir, 'quality-plot.html')\n",
        "        ov = os.path.join(tmpdir, 'overview.html')\n",
        "        dfs = pandas.read_html(fp, index_col=0)\n",
        "        df2s = pandas.read_html(ov, index_col=0)\n",
        "    return dfs + df2s\n",
        "\n",
        "def get_tax(tax: str) -> str:\n",
        "    if not ';' in tax:\n",
        "        return tax\n",
        "    tax = tax.split(';')\n",
        "    for t in tax[::-1]:\n",
        "        if ' ' == t[0]:\n",
        "            if 'uncultured' in t:\n",
        "                continue\n",
        "            return t.strip().replace(';', '')\n",
        "        if t[-1:-3] == '_s':\n",
        "            continue\n",
        "        return t"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### Jellyfish microbiome analysis\n",
        "\n",
        "#### Sample background\n"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### Importing the data\n",
        "\n",
        "The fastq files contain single-end reads, sequenced with the 454 GS FLX+ via a pyrosequencing approach\n",
        "\n",
        "```{bash}\n",
        "qiime tools import \\\n",
        "    --type 'SampleData[SequencesWithQuality]' \\\n",
        "    --input-path data.tsv \\\n",
        "    --output-path jelly.qza \\\n",
        "    --input-format SingleEndFastqManifestPhred33V2\n",
        "```\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [],
      "source": [
        "raw_data = Artifact.load('artifacts/jelly.qza')"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### Raw data exploration\n",
        "\n",
        "The initial data exploration was performed with fastqc. There is a high variation in read length, though the majority are in the 450 - 500 bp range. As expected for short-read sequencing, read quality drops at higher read lengths, necessitating trimming. Surprisingly, the sample has little adapter content.\n",
        "\n",
        "#### Blasting overrepresenting sequences\n",
        "\n",
        "To get a rough idea of what microorganisms were represented most in the sample, I used data from `fastqc`'s overrepresented sequences module.\n",
        "The sequences listed in the module were combined, clustered together to remove redundant reads with `cd-hit`, then blasted against the custom 16s rRNA database.\n",
        "\n",
        "- `cd-hit` grouped 238 overrepresented sequences from the 7 samples into only 17 clusters\n",
        "- The following describes the best 3 hits for the top 3 most overrepresented sequences\n",
        "  - 1. Entoplasma, Mesoplasma, Lebetimonas\n",
        "  - 2. _Ferruginivarius sediminum_, Azospirillum, Desulfosporosinus\n",
        "  - 3. Flavobacteria, Dokdonia, Joostella\n"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### Quality control\n",
        "The dada2 plugin denoises sequences and dereplicates them \n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
            "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
            "\n",
            "Command: run_dada.R --input_directory /tmp/qiime2/sc31/data/54baf4e3-b7ca-4f9c-a76c-8897cbea851c/data --output_path /tmp/tmpvop5ny4s/output.tsv.biom --output_track /tmp/tmpvop5ny4s/track.tsv --filtered_directory /tmp/tmpvop5ny4s --truncation_length 600 --trim_left 0 --max_expected_errors 2.0 --truncation_quality_score 2 --max_length Inf --pooling_method independent --chimera_method consensus --min_parental_fold 1.0 --allow_one_off False --num_threads 1 --learn_min_reads 250000 --homopolymer_gap_penalty 1 --band_size 32\n",
            "\n",
            "R version 4.2.2 (2022-10-31) \n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Loading required package: Rcpp\n"
          ]
        },
        {
          "ename": "KeyboardInterrupt",
          "evalue": "",
          "output_type": "error",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
            "Cell \u001b[0;32mIn[12], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m denoised_Ftable, denoised_Seqs, denoised_stats \u001b[39m=\u001b[39m denoise_pyro(raw_data, trunc_len\u001b[39m=\u001b[39;49m\u001b[39m600\u001b[39;49m) \u001b[39m# Will begin filtering at 600 bp\u001b[39;00m\n\u001b[1;32m      2\u001b[0m denoised_Ftable\u001b[39m.\u001b[39msave(\u001b[39m'\u001b[39m\u001b[39martifacts/denoised.qza\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[1;32m      3\u001b[0m denoised_Seqs\u001b[39m.\u001b[39msave(\u001b[39m'\u001b[39m\u001b[39martifacts/denoised_seqs.qza\u001b[39m\u001b[39m'\u001b[39m)\n",
            "File \u001b[0;32m<decorator-gen-58>:2\u001b[0m, in \u001b[0;36mdenoise_pyro\u001b[0;34m(demultiplexed_seqs, trunc_len, trim_left, max_ee, trunc_q, max_len, pooling_method, chimera_method, min_fold_parent_over_abundance, allow_one_off, n_threads, n_reads_learn, hashed_feature_ids)\u001b[0m\n",
            "File \u001b[0;32m~/Bio_SDD/miniconda3/envs/qiime2-2023.2/lib/python3.8/site-packages/qiime2/sdk/action.py:234\u001b[0m, in \u001b[0;36mAction._bind.<locals>.bound_callable\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    230\u001b[0m         warn(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_build_deprecation_message(),\n\u001b[1;32m    231\u001b[0m              \u001b[39mFutureWarning\u001b[39;00m)\n\u001b[1;32m    233\u001b[0m \u001b[39m# Execute\u001b[39;00m\n\u001b[0;32m--> 234\u001b[0m outputs \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_callable_executor_(scope, callable_args,\n\u001b[1;32m    235\u001b[0m                                    output_types, provenance)\n\u001b[1;32m    237\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(outputs) \u001b[39m!=\u001b[39m \u001b[39mlen\u001b[39m(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39msignature\u001b[39m.\u001b[39moutputs):\n\u001b[1;32m    238\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\n\u001b[1;32m    239\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mNumber of callable outputs must match number of \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    240\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39moutputs defined in signature: \u001b[39m\u001b[39m%d\u001b[39;00m\u001b[39m != \u001b[39m\u001b[39m%d\u001b[39;00m\u001b[39m\"\u001b[39m \u001b[39m%\u001b[39m\n\u001b[1;32m    241\u001b[0m         (\u001b[39mlen\u001b[39m(outputs), \u001b[39mlen\u001b[39m(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39msignature\u001b[39m.\u001b[39moutputs)))\n",
            "File \u001b[0;32m~/Bio_SDD/miniconda3/envs/qiime2-2023.2/lib/python3.8/site-packages/qiime2/sdk/action.py:381\u001b[0m, in \u001b[0;36mMethod._callable_executor_\u001b[0;34m(self, scope, view_args, output_types, provenance)\u001b[0m\n\u001b[1;32m    380\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_callable_executor_\u001b[39m(\u001b[39mself\u001b[39m, scope, view_args, output_types, provenance):\n\u001b[0;32m--> 381\u001b[0m     output_views \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_callable(\u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mview_args)\n\u001b[1;32m    382\u001b[0m     output_views \u001b[39m=\u001b[39m tuplize(output_views)\n\u001b[1;32m    384\u001b[0m     \u001b[39m# TODO this won't work if the user has annotated their \"view API\" to\u001b[39;00m\n\u001b[1;32m    385\u001b[0m     \u001b[39m# return a `typing.Tuple` with some number of components. Python will\u001b[39;00m\n\u001b[1;32m    386\u001b[0m     \u001b[39m# return a tuple when there are multiple return values, and this length\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    389\u001b[0m     \u001b[39m# due to how Python handles multiple returns, and can be worked around\u001b[39;00m\n\u001b[1;32m    390\u001b[0m     \u001b[39m# by using something like `typing.List` instead.\u001b[39;00m\n",
            "File \u001b[0;32m~/Bio_SDD/miniconda3/envs/qiime2-2023.2/lib/python3.8/site-packages/q2_dada2/_denoise.py:367\u001b[0m, in \u001b[0;36mdenoise_pyro\u001b[0;34m(demultiplexed_seqs, trunc_len, trim_left, max_ee, trunc_q, max_len, pooling_method, chimera_method, min_fold_parent_over_abundance, allow_one_off, n_threads, n_reads_learn, hashed_feature_ids)\u001b[0m\n\u001b[1;32m    357\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mdenoise_pyro\u001b[39m(demultiplexed_seqs: SingleLanePerSampleSingleEndFastqDirFmt,\n\u001b[1;32m    358\u001b[0m                  trunc_len: \u001b[39mint\u001b[39m, trim_left: \u001b[39mint\u001b[39m \u001b[39m=\u001b[39m \u001b[39m0\u001b[39m, max_ee: \u001b[39mfloat\u001b[39m \u001b[39m=\u001b[39m \u001b[39m2.0\u001b[39m,\n\u001b[1;32m    359\u001b[0m                  trunc_q: \u001b[39mint\u001b[39m \u001b[39m=\u001b[39m \u001b[39m2\u001b[39m, max_len: \u001b[39mint\u001b[39m \u001b[39m=\u001b[39m \u001b[39m0\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    365\u001b[0m                  hashed_feature_ids: \u001b[39mbool\u001b[39m \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n\u001b[1;32m    366\u001b[0m                  ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m (biom\u001b[39m.\u001b[39mTable, DNAIterator, qiime2\u001b[39m.\u001b[39mMetadata):\n\u001b[0;32m--> 367\u001b[0m     \u001b[39mreturn\u001b[39;00m _denoise_single(\n\u001b[1;32m    368\u001b[0m         demultiplexed_seqs\u001b[39m=\u001b[39;49mdemultiplexed_seqs,\n\u001b[1;32m    369\u001b[0m         trunc_len\u001b[39m=\u001b[39;49mtrunc_len,\n\u001b[1;32m    370\u001b[0m         trim_left\u001b[39m=\u001b[39;49mtrim_left,\n\u001b[1;32m    371\u001b[0m         max_ee\u001b[39m=\u001b[39;49mmax_ee,\n\u001b[1;32m    372\u001b[0m         trunc_q\u001b[39m=\u001b[39;49mtrunc_q,\n\u001b[1;32m    373\u001b[0m         max_len\u001b[39m=\u001b[39;49mmax_len,\n\u001b[1;32m    374\u001b[0m         pooling_method\u001b[39m=\u001b[39;49mpooling_method,\n\u001b[1;32m    375\u001b[0m         chimera_method\u001b[39m=\u001b[39;49mchimera_method,\n\u001b[1;32m    376\u001b[0m         min_fold_parent_over_abundance\u001b[39m=\u001b[39;49mmin_fold_parent_over_abundance,\n\u001b[1;32m    377\u001b[0m         allow_one_off\u001b[39m=\u001b[39;49mallow_one_off,\n\u001b[1;32m    378\u001b[0m         n_threads\u001b[39m=\u001b[39;49mn_threads,\n\u001b[1;32m    379\u001b[0m         n_reads_learn\u001b[39m=\u001b[39;49mn_reads_learn,\n\u001b[1;32m    380\u001b[0m         hashed_feature_ids\u001b[39m=\u001b[39;49mhashed_feature_ids,\n\u001b[1;32m    381\u001b[0m         homopolymer_gap_penalty\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39m1\u001b[39;49m\u001b[39m'\u001b[39;49m,\n\u001b[1;32m    382\u001b[0m         band_size\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39m32\u001b[39;49m\u001b[39m'\u001b[39;49m)\n",
            "File \u001b[0;32m~/Bio_SDD/miniconda3/envs/qiime2-2023.2/lib/python3.8/site-packages/q2_dada2/_denoise.py:220\u001b[0m, in \u001b[0;36m_denoise_single\u001b[0;34m(demultiplexed_seqs, trunc_len, trim_left, max_ee, trunc_q, max_len, pooling_method, chimera_method, min_fold_parent_over_abundance, allow_one_off, n_threads, n_reads_learn, hashed_feature_ids, homopolymer_gap_penalty, band_size)\u001b[0m\n\u001b[1;32m    201\u001b[0m cmd \u001b[39m=\u001b[39m [\u001b[39m'\u001b[39m\u001b[39mrun_dada.R\u001b[39m\u001b[39m'\u001b[39m,\n\u001b[1;32m    202\u001b[0m        \u001b[39m'\u001b[39m\u001b[39m--input_directory\u001b[39m\u001b[39m'\u001b[39m, \u001b[39mstr\u001b[39m(demultiplexed_seqs),\n\u001b[1;32m    203\u001b[0m        \u001b[39m'\u001b[39m\u001b[39m--output_path\u001b[39m\u001b[39m'\u001b[39m, biom_fp,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    217\u001b[0m        \u001b[39m'\u001b[39m\u001b[39m--homopolymer_gap_penalty\u001b[39m\u001b[39m'\u001b[39m, \u001b[39mstr\u001b[39m(homopolymer_gap_penalty),\n\u001b[1;32m    218\u001b[0m        \u001b[39m'\u001b[39m\u001b[39m--band_size\u001b[39m\u001b[39m'\u001b[39m, \u001b[39mstr\u001b[39m(band_size)]\n\u001b[1;32m    219\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 220\u001b[0m     run_commands([cmd])\n\u001b[1;32m    221\u001b[0m \u001b[39mexcept\u001b[39;00m subprocess\u001b[39m.\u001b[39mCalledProcessError \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    222\u001b[0m     \u001b[39mif\u001b[39;00m e\u001b[39m.\u001b[39mreturncode \u001b[39m==\u001b[39m \u001b[39m2\u001b[39m:\n",
            "File \u001b[0;32m~/Bio_SDD/miniconda3/envs/qiime2-2023.2/lib/python3.8/site-packages/q2_dada2/_denoise.py:36\u001b[0m, in \u001b[0;36mrun_commands\u001b[0;34m(cmds, verbose)\u001b[0m\n\u001b[1;32m     34\u001b[0m     \u001b[39mprint\u001b[39m(\u001b[39m\"\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39mCommand:\u001b[39m\u001b[39m\"\u001b[39m, end\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39m \u001b[39m\u001b[39m'\u001b[39m)\n\u001b[1;32m     35\u001b[0m     \u001b[39mprint\u001b[39m(\u001b[39m\"\u001b[39m\u001b[39m \u001b[39m\u001b[39m\"\u001b[39m\u001b[39m.\u001b[39mjoin(cmd), end\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m\\n\u001b[39;00m\u001b[39m'\u001b[39m)\n\u001b[0;32m---> 36\u001b[0m subprocess\u001b[39m.\u001b[39;49mrun(cmd, check\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m)\n",
            "File \u001b[0;32m~/Bio_SDD/miniconda3/envs/qiime2-2023.2/lib/python3.8/subprocess.py:495\u001b[0m, in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    493\u001b[0m \u001b[39mwith\u001b[39;00m Popen(\u001b[39m*\u001b[39mpopenargs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs) \u001b[39mas\u001b[39;00m process:\n\u001b[1;32m    494\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 495\u001b[0m         stdout, stderr \u001b[39m=\u001b[39m process\u001b[39m.\u001b[39;49mcommunicate(\u001b[39minput\u001b[39;49m, timeout\u001b[39m=\u001b[39;49mtimeout)\n\u001b[1;32m    496\u001b[0m     \u001b[39mexcept\u001b[39;00m TimeoutExpired \u001b[39mas\u001b[39;00m exc:\n\u001b[1;32m    497\u001b[0m         process\u001b[39m.\u001b[39mkill()\n",
            "File \u001b[0;32m~/Bio_SDD/miniconda3/envs/qiime2-2023.2/lib/python3.8/subprocess.py:1020\u001b[0m, in \u001b[0;36mPopen.communicate\u001b[0;34m(self, input, timeout)\u001b[0m\n\u001b[1;32m   1018\u001b[0m         stderr \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mstderr\u001b[39m.\u001b[39mread()\n\u001b[1;32m   1019\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mstderr\u001b[39m.\u001b[39mclose()\n\u001b[0;32m-> 1020\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mwait()\n\u001b[1;32m   1021\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m   1022\u001b[0m     \u001b[39mif\u001b[39;00m timeout \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
            "File \u001b[0;32m~/Bio_SDD/miniconda3/envs/qiime2-2023.2/lib/python3.8/subprocess.py:1083\u001b[0m, in \u001b[0;36mPopen.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1081\u001b[0m     endtime \u001b[39m=\u001b[39m _time() \u001b[39m+\u001b[39m timeout\n\u001b[1;32m   1082\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1083\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_wait(timeout\u001b[39m=\u001b[39;49mtimeout)\n\u001b[1;32m   1084\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyboardInterrupt\u001b[39;00m:\n\u001b[1;32m   1085\u001b[0m     \u001b[39m# https://bugs.python.org/issue25942\u001b[39;00m\n\u001b[1;32m   1086\u001b[0m     \u001b[39m# The first keyboard interrupt waits briefly for the child to\u001b[39;00m\n\u001b[1;32m   1087\u001b[0m     \u001b[39m# exit under the common assumption that it also received the ^C\u001b[39;00m\n\u001b[1;32m   1088\u001b[0m     \u001b[39m# generated SIGINT and will exit rapidly.\u001b[39;00m\n\u001b[1;32m   1089\u001b[0m     \u001b[39mif\u001b[39;00m timeout \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
            "File \u001b[0;32m~/Bio_SDD/miniconda3/envs/qiime2-2023.2/lib/python3.8/subprocess.py:1806\u001b[0m, in \u001b[0;36mPopen._wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1804\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mreturncode \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m   1805\u001b[0m     \u001b[39mbreak\u001b[39;00m  \u001b[39m# Another thread waited.\u001b[39;00m\n\u001b[0;32m-> 1806\u001b[0m (pid, sts) \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_try_wait(\u001b[39m0\u001b[39;49m)\n\u001b[1;32m   1807\u001b[0m \u001b[39m# Check the pid and loop as waitpid has been known to\u001b[39;00m\n\u001b[1;32m   1808\u001b[0m \u001b[39m# return 0 even without WNOHANG in odd situations.\u001b[39;00m\n\u001b[1;32m   1809\u001b[0m \u001b[39m# http://bugs.python.org/issue14396.\u001b[39;00m\n\u001b[1;32m   1810\u001b[0m \u001b[39mif\u001b[39;00m pid \u001b[39m==\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mpid:\n",
            "File \u001b[0;32m~/Bio_SDD/miniconda3/envs/qiime2-2023.2/lib/python3.8/subprocess.py:1764\u001b[0m, in \u001b[0;36mPopen._try_wait\u001b[0;34m(self, wait_flags)\u001b[0m\n\u001b[1;32m   1762\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"All callers to this function MUST hold self._waitpid_lock.\"\"\"\u001b[39;00m\n\u001b[1;32m   1763\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1764\u001b[0m     (pid, sts) \u001b[39m=\u001b[39m os\u001b[39m.\u001b[39;49mwaitpid(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mpid, wait_flags)\n\u001b[1;32m   1765\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mChildProcessError\u001b[39;00m:\n\u001b[1;32m   1766\u001b[0m     \u001b[39m# This happens if SIGCLD is set to be ignored or waiting\u001b[39;00m\n\u001b[1;32m   1767\u001b[0m     \u001b[39m# for child processes has otherwise been disabled for our\u001b[39;00m\n\u001b[1;32m   1768\u001b[0m     \u001b[39m# process.  This child is dead, we can't get the status.\u001b[39;00m\n\u001b[1;32m   1769\u001b[0m     pid \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mpid\n",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
          ]
        }
      ],
      "source": [
        "denoised_Ftable, denoised_Seqs, denoised_stats = denoise_pyro(raw_data, trunc_len=600) # Will begin filtering at 600 bp\n",
        "denoised_Ftable.save('artifacts/denoised.qza')\n",
        "denoised_Seqs.save('artifacts/denoised_seqs.qza')\n",
        "stats_viz, = metadata_actions.tabulate(input=denoised_stats.view(Metadata))\n",
        "stats_viz.save('vis/denoised.qzv')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "FeatureData[Sequence]\n",
            "FeatureTable[Frequency]\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "'vis/sequences.qzv'"
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Importing previous\n",
        "denoised_FtableDF = denoised_Ftable.view(pd.DataFrame).T\n",
        "denoised_FtableDF.index.names = ['Feature ID']\n",
        "denoised_FtableMD = denoised_Ftable.view(Metadata)\n",
        "seqs_viz, = fta.tabulate_seqs(data=denoised_Seqs) # Tabulate sequences to \n",
        "denoised_SeqsDF = denoised_Seqs.view(Metadata).to_dataframe()\n",
        "seqs_viz.save('vis/sequences.qzv')"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "##### Clustering\n",
        "Once the sequences have been dereplicated, they can be clustered them into operational taxonomic units (OTUs). Given that jellyfish microorganisms may not be as well represented in databases, I'll be using de novo clustering rather than reference-based"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 28,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Running external command line application. This may print messages to stdout and/or stderr.\n",
            "The command being run is below. This command cannot be manually re-run as it will depend on temporary files that no longer exist.\n",
            "\n",
            "Command: vsearch --cluster_size /tmp/tmpgqn7og7s --id 0.99 --centroids /tmp/q2-DNAFASTAFormat-82pdf45f --uc /tmp/tmp9mboselc --qmask none --xsize --threads 1 --minseqlength 1 --fasta_width 0\n",
            "\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "vsearch v2.22.1_linux_x86_64, 22.9GB RAM, 16 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading file /tmp/tmpgqn7og7s 100%\n",
            "15600 nt in 26 seqs, min 600, max 600, avg 600\n",
            "Sorting by abundance 100%\n",
            "Counting k-mers 100%\n",
            "Clustering 100%\n",
            "Sorting clusters 100%\n",
            "Writing clusters 100%\n",
            "Clusters: 20 Size min 1, max 3, avg 1.3\n",
            "Singletons: 16, 61.5% of seqs, 80.0% of clusters\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "'artifacts/otuSeqs.qza'"
            ]
          },
          "execution_count": 28,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "otuFreqs, otuSeqs = vsm.cluster_features_de_novo(denoised_Seqs, denoised_Ftable, 0.99)\n",
        "otuFreqs.save('artifacts/otuFreqs.qza')\n",
        "otuSeqs.save('artifacts/otuSeqs.qza')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 33,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Sequence</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Feature ID</th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>8331adf13c184313c464be9c7b21097c</th>\n",
              "      <td>CACTCTTGCGAGCATACTACTCAGGCGGAGTACTTAACGCGTTAGC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>b5f4256d637420120c2e71bc85666bfd</th>\n",
              "      <td>CATTCTTGCGAACGTACTCCCCAGGTGGGATACTTATCACTTTCGC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>00aabc54895675bbc65b0c1fdaf1fd07</th>\n",
              "      <td>TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>fe008f8f1f5a1776177e14bae97e10e2</th>\n",
              "      <td>CACACTTGCGTGCGTACTCCCCAGGCGGAACACTTAACGCGTTGGC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3a4c4c7d2b8e5a489eb2d1fddc95e025</th>\n",
              "      <td>TAATCTTGCGACCGTACTCCCCAGGCGGAATGCTTAATCCGTTAGG...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6d4121b0c9e7eda284b4ab5bf571dc86</th>\n",
              "      <td>TAGTCTTGCGACCGTAGTCCCCAGGCGGAGTGCTTAACGCGTTAGC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>c6eddca0dc7ee0b43403e873ba6c8cfc</th>\n",
              "      <td>TAGTCTTGCGACCGTACTCCCCAGGCGGAGAACTTAACGCGTTAGC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>d4c1d85ff4ca5efdc346801641b2d9a2</th>\n",
              "      <td>TAATCTTGCGACCGTACTCCCCAGGCGGTCTACTTAACGCGTTAGC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>c00ee893109657877cc78cca52003ca9</th>\n",
              "      <td>CAGCCTTGCGACCATACTCCCCAGGCGGAACACTTAACGCTTTCGC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>d6f1d18cadc754defc476dd14e4f4c58</th>\n",
              "      <td>TAATCTTGCGACCGTACTCCCCAGGCGGTTCATTTAATGCGTTAGC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4d0e9bcbdf46ba60e8c807aca0550f26</th>\n",
              "      <td>TAACCTTGCGGCCGTACTCCCCAGGCGGTCAACTTATCGCGTTAGC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3046631d2ae55b33a73e710aecb4db9e</th>\n",
              "      <td>TAGCCTTGCGGCCGTACTCCCCAGGCGGTACACTTATCGCGTTAGC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6ae71cd37d021784c75d9c4ca7de333d</th>\n",
              "      <td>TAATCTTGCGACCGTACTCCCCAGGCGGAGTGCTTAATGCGTTAGC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>a6c50d7fef8a40824552274d0e96e3c4</th>\n",
              "      <td>TAGCCTTGCGGCCGTACTCCCCAGGCGGATAACTTAACGCGTTAGC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>c27592c0acdd2708a5b18cf790d0d1e1</th>\n",
              "      <td>CAACCTTGCGGCCGTACTCCCCAGGCGGAGTGCTTAATGCGTTAGC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>f67e6cc781a78a69a56d9d441411694b</th>\n",
              "      <td>TAGCCTTGCGGCCGTACTCCCCAGGCGGACAACTTAAAACGTTAGC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>b770c9ce6a8ca4de3d7085854747551c</th>\n",
              "      <td>TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>c74f9ca1eae5081f61f84934c8e22a79</th>\n",
              "      <td>TAATCTTGCGACCGTACTCCCCAGGCGGTCTACTTATCGCGTTAGC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>c92bce272a01f65e7105823ab7957ae8</th>\n",
              "      <td>CATTCTTGCGAACGTACTCCCCAGGTGGGATACTTATCACTTTCGC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>f1a3e9623a42e1ee02b2317bb9a02a25</th>\n",
              "      <td>TAGCCTTGCGGCCGTACTCCCCAGGCGGAGAACTTATCGCGTTAGC...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                                                           Sequence\n",
              "Feature ID                                                                         \n",
              "8331adf13c184313c464be9c7b21097c  CACTCTTGCGAGCATACTACTCAGGCGGAGTACTTAACGCGTTAGC...\n",
              "b5f4256d637420120c2e71bc85666bfd  CATTCTTGCGAACGTACTCCCCAGGTGGGATACTTATCACTTTCGC...\n",
              "00aabc54895675bbc65b0c1fdaf1fd07  TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...\n",
              "fe008f8f1f5a1776177e14bae97e10e2  CACACTTGCGTGCGTACTCCCCAGGCGGAACACTTAACGCGTTGGC...\n",
              "3a4c4c7d2b8e5a489eb2d1fddc95e025  TAATCTTGCGACCGTACTCCCCAGGCGGAATGCTTAATCCGTTAGG...\n",
              "6d4121b0c9e7eda284b4ab5bf571dc86  TAGTCTTGCGACCGTAGTCCCCAGGCGGAGTGCTTAACGCGTTAGC...\n",
              "c6eddca0dc7ee0b43403e873ba6c8cfc  TAGTCTTGCGACCGTACTCCCCAGGCGGAGAACTTAACGCGTTAGC...\n",
              "d4c1d85ff4ca5efdc346801641b2d9a2  TAATCTTGCGACCGTACTCCCCAGGCGGTCTACTTAACGCGTTAGC...\n",
              "c00ee893109657877cc78cca52003ca9  CAGCCTTGCGACCATACTCCCCAGGCGGAACACTTAACGCTTTCGC...\n",
              "d6f1d18cadc754defc476dd14e4f4c58  TAATCTTGCGACCGTACTCCCCAGGCGGTTCATTTAATGCGTTAGC...\n",
              "4d0e9bcbdf46ba60e8c807aca0550f26  TAACCTTGCGGCCGTACTCCCCAGGCGGTCAACTTATCGCGTTAGC...\n",
              "3046631d2ae55b33a73e710aecb4db9e  TAGCCTTGCGGCCGTACTCCCCAGGCGGTACACTTATCGCGTTAGC...\n",
              "6ae71cd37d021784c75d9c4ca7de333d  TAATCTTGCGACCGTACTCCCCAGGCGGAGTGCTTAATGCGTTAGC...\n",
              "a6c50d7fef8a40824552274d0e96e3c4  TAGCCTTGCGGCCGTACTCCCCAGGCGGATAACTTAACGCGTTAGC...\n",
              "c27592c0acdd2708a5b18cf790d0d1e1  CAACCTTGCGGCCGTACTCCCCAGGCGGAGTGCTTAATGCGTTAGC...\n",
              "f67e6cc781a78a69a56d9d441411694b  TAGCCTTGCGGCCGTACTCCCCAGGCGGACAACTTAAAACGTTAGC...\n",
              "b770c9ce6a8ca4de3d7085854747551c  TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...\n",
              "c74f9ca1eae5081f61f84934c8e22a79  TAATCTTGCGACCGTACTCCCCAGGCGGTCTACTTATCGCGTTAGC...\n",
              "c92bce272a01f65e7105823ab7957ae8  CATTCTTGCGAACGTACTCCCCAGGTGGGATACTTATCACTTTCGC...\n",
              "f1a3e9623a42e1ee02b2317bb9a02a25  TAGCCTTGCGGCCGTACTCCCCAGGCGGAGAACTTATCGCGTTAGC..."
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "otuSeqsDF = otuSeqs.view(Metadata).to_dataframe()"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### Pairing sequences with their frequencies in Pandas\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>sample-1</th>\n",
              "      <th>sample-2</th>\n",
              "      <th>sample-3</th>\n",
              "      <th>sample-4</th>\n",
              "      <th>sample-5</th>\n",
              "      <th>sample-6</th>\n",
              "      <th>sample-7</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Feature ID</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>92.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>5247.0</td>\n",
              "      <td>1983.0</td>\n",
              "      <td>CACTCTTGCGAGCATACTACTCAGGCGGAGTACTTAACGCGTTAGC...</td>\n",
              "      <td>8331adf13c184313c464be9c7b21097c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>4268.0</td>\n",
              "      <td>2209.0</td>\n",
              "      <td>325.0</td>\n",
              "      <td>392.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>CACTCTTGCGAGCATACTACTCAGGCGGAGTACTTAACGCGTTAGC...</td>\n",
              "      <td>b37b36ef2cd58aa19929a8ae4f8f4c21</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>671.0</td>\n",
              "      <td>277.0</td>\n",
              "      <td>97.0</td>\n",
              "      <td>30.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>387.0</td>\n",
              "      <td>42.0</td>\n",
              "      <td>CATTCTTGCGAACGTACTCCCCAGGTGGGATACTTATCACTTTCGC...</td>\n",
              "      <td>b5f4256d637420120c2e71bc85666bfd</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>94.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>298.0</td>\n",
              "      <td>284.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>6.0</td>\n",
              "      <td>TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...</td>\n",
              "      <td>00aabc54895675bbc65b0c1fdaf1fd07</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>681.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>CACTCTTGCGAGCATACTACTCAGGCGGAGTACTTAACGCGTTAGC...</td>\n",
              "      <td>3c5ee7863e02dd5604419d803a014014</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>235.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>121.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>CACACTTGCGTGCGTACTCCCCAGGCGGAACACTTAACGCGTTGGC...</td>\n",
              "      <td>fe008f8f1f5a1776177e14bae97e10e2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>188.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAATCTTGCGACCGTACTCCCCAGGCGGAATGCTTAATCCGTTAGG...</td>\n",
              "      <td>3a4c4c7d2b8e5a489eb2d1fddc95e025</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>163.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...</td>\n",
              "      <td>0c2db32fc52f027a04e433a93c0669df</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>76.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...</td>\n",
              "      <td>5f460fa8105db752feb6166a52b135d2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>45.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAGTCTTGCGACCGTAGTCCCCAGGCGGAGTGCTTAACGCGTTAGC...</td>\n",
              "      <td>6d4121b0c9e7eda284b4ab5bf571dc86</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>16.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>CACACTTGCGTGCGTACTCCCCAGGCGGAACACTTAACGCGTTGGC...</td>\n",
              "      <td>6e0740a41b665e09ed40342b621fb3b0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>13.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAGTCTTGCGACCGTACTCCCCAGGCGGAGAACTTAACGCGTTAGC...</td>\n",
              "      <td>c6eddca0dc7ee0b43403e873ba6c8cfc</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>11.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAATCTTGCGACCGTACTCCCCAGGCGGTCTACTTAACGCGTTAGC...</td>\n",
              "      <td>d4c1d85ff4ca5efdc346801641b2d9a2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>10.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>CAGCCTTGCGACCATACTCCCCAGGCGGAACACTTAACGCTTTCGC...</td>\n",
              "      <td>c00ee893109657877cc78cca52003ca9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>9.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAATCTTGCGACCGTACTCCCCAGGCGGTTCATTTAATGCGTTAGC...</td>\n",
              "      <td>d6f1d18cadc754defc476dd14e4f4c58</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>8.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>CATTCTTGCGAACGTACTCCCCAGGTGGGATACTTATCACTTTCGC...</td>\n",
              "      <td>ac30e74e21dd088eabfea4d8738c6279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>7.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAACCTTGCGGCCGTACTCCCCAGGCGGTCAACTTATCGCGTTAGC...</td>\n",
              "      <td>4d0e9bcbdf46ba60e8c807aca0550f26</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>5.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAGCCTTGCGGCCGTACTCCCCAGGCGGTACACTTATCGCGTTAGC...</td>\n",
              "      <td>3046631d2ae55b33a73e710aecb4db9e</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>0.0</td>\n",
              "      <td>5.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAATCTTGCGACCGTACTCCCCAGGCGGAGTGCTTAATGCGTTAGC...</td>\n",
              "      <td>6ae71cd37d021784c75d9c4ca7de333d</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>4.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>CAACCTTGCGGCCGTACTCCCCAGGCGGAGTGCTTAATGCGTTAGC...</td>\n",
              "      <td>c27592c0acdd2708a5b18cf790d0d1e1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>4.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAGCCTTGCGGCCGTACTCCCCAGGCGGATAACTTAACGCGTTAGC...</td>\n",
              "      <td>a6c50d7fef8a40824552274d0e96e3c4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>4.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAGCCTTGCGGCCGTACTCCCCAGGCGGACAACTTAAAACGTTAGC...</td>\n",
              "      <td>f67e6cc781a78a69a56d9d441411694b</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>3.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>CATTCTTGCGAACGTACTCCCCAGGTGGGATACTTATCACTTTCGC...</td>\n",
              "      <td>c92bce272a01f65e7105823ab7957ae8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>3.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAATCTTGCGACCGTACTCCCCAGGCGGTCTACTTATCGCGTTAGC...</td>\n",
              "      <td>c74f9ca1eae5081f61f84934c8e22a79</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>3.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...</td>\n",
              "      <td>b770c9ce6a8ca4de3d7085854747551c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>2.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAGCCTTGCGGCCGTACTCCCCAGGCGGAGAACTTATCGCGTTAGC...</td>\n",
              "      <td>f1a3e9623a42e1ee02b2317bb9a02a25</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "    sample-1  sample-2  sample-3  sample-4  sample-5  sample-6  sample-7  \\\n",
              "0        0.0       0.0      92.0       0.0       0.0    5247.0    1983.0   \n",
              "1     4268.0    2209.0     325.0     392.0       0.0       0.0       0.0   \n",
              "2      671.0     277.0      97.0      30.0       0.0     387.0      42.0   \n",
              "3       94.0       0.0     298.0     284.0       0.0       0.0       6.0   \n",
              "4        0.0       0.0       0.0       0.0     681.0       0.0       0.0   \n",
              "5      235.0       0.0       0.0       0.0       0.0     121.0       0.0   \n",
              "6      188.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
              "7        0.0       0.0       0.0       0.0       0.0     163.0       0.0   \n",
              "8        0.0       0.0       0.0       0.0      76.0       0.0       0.0   \n",
              "9        0.0       0.0       0.0       0.0       0.0      45.0       0.0   \n",
              "10       0.0       0.0      16.0       0.0       0.0       0.0       0.0   \n",
              "11       0.0       0.0       0.0       0.0       0.0      13.0       0.0   \n",
              "12      11.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
              "13      10.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
              "14       0.0       0.0       0.0       0.0       0.0       9.0       0.0   \n",
              "15       0.0       0.0       0.0       0.0       8.0       0.0       0.0   \n",
              "16       0.0       0.0       0.0       0.0       0.0       7.0       0.0   \n",
              "17       5.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
              "18       0.0       5.0       0.0       0.0       0.0       0.0       0.0   \n",
              "19       4.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
              "20       4.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
              "21       0.0       0.0       4.0       0.0       0.0       0.0       0.0   \n",
              "22       3.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
              "23       3.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
              "24       0.0       0.0       0.0       3.0       0.0       0.0       0.0   \n",
              "25       0.0       0.0       0.0       0.0       0.0       2.0       0.0   \n",
              "\n",
              "                                             Sequence  \\\n",
              "0   CACTCTTGCGAGCATACTACTCAGGCGGAGTACTTAACGCGTTAGC...   \n",
              "1   CACTCTTGCGAGCATACTACTCAGGCGGAGTACTTAACGCGTTAGC...   \n",
              "2   CATTCTTGCGAACGTACTCCCCAGGTGGGATACTTATCACTTTCGC...   \n",
              "3   TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...   \n",
              "4   CACTCTTGCGAGCATACTACTCAGGCGGAGTACTTAACGCGTTAGC...   \n",
              "5   CACACTTGCGTGCGTACTCCCCAGGCGGAACACTTAACGCGTTGGC...   \n",
              "6   TAATCTTGCGACCGTACTCCCCAGGCGGAATGCTTAATCCGTTAGG...   \n",
              "7   TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...   \n",
              "8   TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...   \n",
              "9   TAGTCTTGCGACCGTAGTCCCCAGGCGGAGTGCTTAACGCGTTAGC...   \n",
              "10  CACACTTGCGTGCGTACTCCCCAGGCGGAACACTTAACGCGTTGGC...   \n",
              "11  TAGTCTTGCGACCGTACTCCCCAGGCGGAGAACTTAACGCGTTAGC...   \n",
              "12  TAATCTTGCGACCGTACTCCCCAGGCGGTCTACTTAACGCGTTAGC...   \n",
              "13  CAGCCTTGCGACCATACTCCCCAGGCGGAACACTTAACGCTTTCGC...   \n",
              "14  TAATCTTGCGACCGTACTCCCCAGGCGGTTCATTTAATGCGTTAGC...   \n",
              "15  CATTCTTGCGAACGTACTCCCCAGGTGGGATACTTATCACTTTCGC...   \n",
              "16  TAACCTTGCGGCCGTACTCCCCAGGCGGTCAACTTATCGCGTTAGC...   \n",
              "17  TAGCCTTGCGGCCGTACTCCCCAGGCGGTACACTTATCGCGTTAGC...   \n",
              "18  TAATCTTGCGACCGTACTCCCCAGGCGGAGTGCTTAATGCGTTAGC...   \n",
              "19  CAACCTTGCGGCCGTACTCCCCAGGCGGAGTGCTTAATGCGTTAGC...   \n",
              "20  TAGCCTTGCGGCCGTACTCCCCAGGCGGATAACTTAACGCGTTAGC...   \n",
              "21  TAGCCTTGCGGCCGTACTCCCCAGGCGGACAACTTAAAACGTTAGC...   \n",
              "22  CATTCTTGCGAACGTACTCCCCAGGTGGGATACTTATCACTTTCGC...   \n",
              "23  TAATCTTGCGACCGTACTCCCCAGGCGGTCTACTTATCGCGTTAGC...   \n",
              "24  TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...   \n",
              "25  TAGCCTTGCGGCCGTACTCCCCAGGCGGAGAACTTATCGCGTTAGC...   \n",
              "\n",
              "                          Feature ID  \n",
              "0   8331adf13c184313c464be9c7b21097c  \n",
              "1   b37b36ef2cd58aa19929a8ae4f8f4c21  \n",
              "2   b5f4256d637420120c2e71bc85666bfd  \n",
              "3   00aabc54895675bbc65b0c1fdaf1fd07  \n",
              "4   3c5ee7863e02dd5604419d803a014014  \n",
              "5   fe008f8f1f5a1776177e14bae97e10e2  \n",
              "6   3a4c4c7d2b8e5a489eb2d1fddc95e025  \n",
              "7   0c2db32fc52f027a04e433a93c0669df  \n",
              "8   5f460fa8105db752feb6166a52b135d2  \n",
              "9   6d4121b0c9e7eda284b4ab5bf571dc86  \n",
              "10  6e0740a41b665e09ed40342b621fb3b0  \n",
              "11  c6eddca0dc7ee0b43403e873ba6c8cfc  \n",
              "12  d4c1d85ff4ca5efdc346801641b2d9a2  \n",
              "13  c00ee893109657877cc78cca52003ca9  \n",
              "14  d6f1d18cadc754defc476dd14e4f4c58  \n",
              "15  ac30e74e21dd088eabfea4d8738c6279  \n",
              "16  4d0e9bcbdf46ba60e8c807aca0550f26  \n",
              "17  3046631d2ae55b33a73e710aecb4db9e  \n",
              "18  6ae71cd37d021784c75d9c4ca7de333d  \n",
              "19  c27592c0acdd2708a5b18cf790d0d1e1  \n",
              "20  a6c50d7fef8a40824552274d0e96e3c4  \n",
              "21  f67e6cc781a78a69a56d9d441411694b  \n",
              "22  c92bce272a01f65e7105823ab7957ae8  \n",
              "23  c74f9ca1eae5081f61f84934c8e22a79  \n",
              "24  b770c9ce6a8ca4de3d7085854747551c  \n",
              "25  f1a3e9623a42e1ee02b2317bb9a02a25  "
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# display(denoised_FtableDF)\n",
        "freq_seq = denoised_FtableDF.merge(denoised_SeqsDF, on='Feature ID')\n",
        "freq_seq['Feature ID'] = freq_seq.index\n",
        "freq_seq = freq_seq.reset_index(drop=True)\n",
        "display(freq_seq)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "<Figure size 640x480 with 2 Axes>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "sns.heatmap(freq_seq.iloc[:, :-2])\n",
        "plt.show()"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### Taxonomic analyses\n"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### Alignment-based classification\n"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "##### Blast\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 34,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Running external command line application. This may print messages to stdout and/or stderr.\n",
            "The command being run is below. This command cannot be manually re-run as it will depend on temporary files that no longer exist.\n",
            "\n",
            "Command: blastn -query /tmp/qiime2/sc31/data/6ce41b78-4959-4a10-ad3a-6fa008e4eb34/data/dna-sequences.fasta -evalue 0.001 -strand both -outfmt 6 -subject /tmp/qiime2/sc31/data/4dca7356-2c27-4284-8679-b3ec0974b1ab/data/dna-sequences.fasta -perc_identity 80.0 -qcov_hsp_perc 80.0 -max_target_seqs 10 -out /tmp/q2-BLAST6Format-24ykq59q\n",
            "\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "'artifacts/blast_hits.qza'"
            ]
          },
          "execution_count": 34,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Run search\n",
        "classified, top_hits = classify_consensus_blast(otuSeqs, ref_seqs, ref_ids)\n",
        "classified.save('artifacts/blast_seqs.qza')\n",
        "top_hits.save('artifacts/blast_hits.qza')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 39,
      "metadata": {},
      "outputs": [],
      "source": [
        "blastDF = blast.view(Metadata).to_dataframe()\n",
        "blastTopDF = blast_top_hits.view(Metadata).to_dataframe()"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "##### Vsearch\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Running external command line application. This may print messages to stdout and/or stderr.\n",
            "The command being run is below. This command cannot be manually re-run as it will depend on temporary files that no longer exist.\n",
            "\n",
            "Command: vsearch --usearch_global /tmp/qiime2/sc31/data/bf89e318-4049-42b4-b410-ebf99b324ef1/data/dna-sequences.fasta --id 0.8 --query_cov 0.8 --strand both --maxaccepts 10 --maxrejects 0 --db /tmp/qiime2/sc31/data/4dca7356-2c27-4284-8679-b3ec0974b1ab/data/dna-sequences.fasta --threads 1 --output_no_hits --blast6out /tmp/q2-BLAST6Format-e2m8jf6c\n",
            "\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "vsearch v2.22.1_linux_x86_64, 22.9GB RAM, 16 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading file /tmp/qiime2/sc31/data/4dca7356-2c27-4284-8679-b3ec0974b1ab/data/dna-sequences.fasta 100%\n",
            "1272167297 nt in 866608 seqs, min 302, max 4563, avg 1468\n",
            "Masking 100%\n",
            "Counting k-mers 100%\n",
            "Creating k-mer index 100%\n",
            "Searching 100%\n",
            "Matching unique query sequences: 26 of 26 (100.00%)\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "'artifacts/vsearch_hits.qza'"
            ]
          },
          "execution_count": 16,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Run search\n",
        "classified, top_hits = classify_consensus_vsearch(denoised_Seqs, ref_seqs, ref_ids)\n",
        "classified.save('artifacts/vsearch_seqs.qza')\n",
        "top_hits.save('artifacts/vsearch_hits.qza')\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {},
      "outputs": [],
      "source": [
        "vsearchDF = vsearch_seqs.view(Metadata).to_dataframe()\n",
        "vsearchTopDF = vsearch_top_hits.view(Metadata).to_dataframe()"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### Machine learning\n"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": []
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Diversity analyses"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### Phylogeny reconstruction\n",
        "- Reconstructing a phylogenetic tree from the otu sequences may allow us to identify unknown otus by placing them in relation to those that were identified. \n",
        "- qiime2 offers several different "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 65,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Running external command line application. This may print messages to stdout and/or stderr.\n",
            "The command being run is below. This command cannot be manually re-run as it will depend on temporary files that no longer exist.\n",
            "\n",
            "Command: mafft --preservecase --inputorder --thread 1 /tmp/qiime2/sc31/data/6ce41b78-4959-4a10-ad3a-6fa008e4eb34/data/dna-sequences.fasta\n",
            "\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "inputfile = orig\n",
            "20 x 600 - 600 d\n",
            "nthread = 1\n",
            "nthreadpair = 1\n",
            "nthreadtb = 1\n",
            "ppenalty_ex = 0\n",
            "stacksize: 8192 kb\n",
            "generating a scoring matrix for nucleotide (dist=200) ... done\n",
            "Gap Penalty = -1.53, +0.00, +0.00\n",
            "\n",
            "\n",
            "\n",
            "Making a distance matrix ..\n",
            "    1 / 20 (thread    0)\n",
            "done.\n",
            "\n",
            "Constructing a UPGMA tree (efffree=0) ... \n",
            "   10 / 20\n",
            "done.\n",
            "\n",
            "Progressive alignment 1/2... \n",
            "STEP     6 / 19 (thread    0)\n",
            "Reallocating..done. *alloclen = 2203\n",
            "STEP    19 / 19 (thread    0)\n",
            "done.\n",
            "\n",
            "Making a distance matrix from msa.. \n",
            "    0 / 20 (thread    0)\n",
            "done.\n",
            "\n",
            "Constructing a UPGMA tree (efffree=1) ... \n",
            "   10 / 20\n",
            "done.\n",
            "\n",
            "Progressive alignment 2/2... \n",
            "STEP     5 / 19 (thread    0)\n",
            "Reallocating..done. *alloclen = 2202\n",
            "STEP    19 / 19 (thread    0)\n",
            "done.\n",
            "\n",
            "disttbfast (nuc) Version 7.515\n",
            "alg=A, model=DNA200 (2), 1.53 (4.59), -0.00 (-0.00), noshift, amax=0.0\n",
            "1 thread(s)\n",
            "\n",
            "\n",
            "Strategy:\n",
            " FFT-NS-2 (Fast but rough)\n",
            " Progressive method (guide trees were built 2 times.)\n",
            "\n",
            "If unsure which option to use, try 'mafft --auto input > output'.\n",
            "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
            "\n",
            "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
            "It tends to insert more gaps into gap-rich regions than previous versions.\n",
            "To disable this change, add the --leavegappyregion option.\n",
            "\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Running external command line application. This may print messages to stdout and/or stderr.\n",
            "The command being run is below. This command cannot be manually re-run as it will depend on temporary files that no longer exist.\n",
            "\n",
            "Command: iqtree -bb 1000 -st DNA --runs 1 -s /tmp/qiime2/sc31/data/3267aa82-df09-4459-92db-2277b184680d/data/aligned-dna-sequences.fasta -m MFP -pre /tmp/tmpdxraaq6_/q2iqtreeufboot -nt 1\n",
            "\n",
            "IQ-TREE multicore version 2.2.0.3 COVID-edition for Linux 64-bit built Aug  2 2022\n",
            "Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,\n",
            "Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams, Ly Trong Nhan.\n",
            "\n",
            "Host:    bravo15-pop (AVX2, FMA3, 22 GB RAM)\n",
            "Command: iqtree -bb 1000 -st DNA --runs 1 -s /tmp/qiime2/sc31/data/3267aa82-df09-4459-92db-2277b184680d/data/aligned-dna-sequences.fasta -m MFP -pre /tmp/tmpdxraaq6_/q2iqtreeufboot -nt 1\n",
            "Seed:    81051 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
            "Time:    Thu Jun  1 06:56:10 2023\n",
            "Kernel:  AVX+FMA - 1 threads (16 CPU cores detected)\n",
            "\n",
            "HINT: Use -nt option to specify number of threads because your CPU has 16 cores!\n",
            "HINT: -nt AUTO will automatically determine the best number of threads to use.\n",
            "\n",
            "Reading alignment file /tmp/qiime2/sc31/data/3267aa82-df09-4459-92db-2277b184680d/data/aligned-dna-sequences.fasta ... Fasta format detected\n",
            "Reading fasta file: done in 0.000178468 secs\n",
            "Alignment most likely contains DNA/RNA sequences\n",
            "Constructing alignment: done in 0.000223152 secs\n",
            "Alignment has 20 sequences with 636 columns, 363 distinct patterns\n",
            "288 parsimony-informative, 69 singleton sites, 279 constant sites\n",
            "                                  Gap/Ambiguity  Composition  p-value\n",
            "Analyzing sequences: done in 1.1983e-05 secs\n",
            "   1  8331adf13c184313c464be9c7b21097c    5.66%    passed      8.40%\n",
            "   2  b5f4256d637420120c2e71bc85666bfd    5.66%    passed     19.48%\n",
            "   3  00aabc54895675bbc65b0c1fdaf1fd07    5.66%    passed     98.18%\n",
            "   4  fe008f8f1f5a1776177e14bae97e10e2    5.66%    passed     21.31%\n",
            "   5  3a4c4c7d2b8e5a489eb2d1fddc95e025    5.66%    passed     91.92%\n",
            "   6  6d4121b0c9e7eda284b4ab5bf571dc86    5.66%    passed     95.53%\n",
            "   7  c6eddca0dc7ee0b43403e873ba6c8cfc    5.66%    passed     82.13%\n",
            "   8  d4c1d85ff4ca5efdc346801641b2d9a2    5.66%    passed     89.45%\n",
            "   9  c00ee893109657877cc78cca52003ca9    5.66%    passed      6.32%\n",
            "  10  d6f1d18cadc754defc476dd14e4f4c58    5.66%    passed     91.75%\n",
            "  11  4d0e9bcbdf46ba60e8c807aca0550f26    5.66%    passed     92.52%\n",
            "  12  3046631d2ae55b33a73e710aecb4db9e    5.66%    passed     73.08%\n",
            "  13  6ae71cd37d021784c75d9c4ca7de333d    5.66%    passed     18.41%\n",
            "  14  a6c50d7fef8a40824552274d0e96e3c4    5.66%    passed     42.35%\n",
            "  15  c27592c0acdd2708a5b18cf790d0d1e1    5.66%    passed      9.36%\n",
            "  16  f67e6cc781a78a69a56d9d441411694b    5.66%    passed     94.20%\n",
            "  17  b770c9ce6a8ca4de3d7085854747551c    5.66%    passed     25.71%\n",
            "  18  c74f9ca1eae5081f61f84934c8e22a79    5.66%    passed     53.41%\n",
            "  19  c92bce272a01f65e7105823ab7957ae8    5.66%    failed      4.11%\n",
            "  20  f1a3e9623a42e1ee02b2317bb9a02a25    5.66%    passed     93.25%\n",
            "****  TOTAL                               5.66%  1 sequences failed composition chi2 test (p-value<5%; df=3)\n",
            "Checking for duplicate sequences: done in 2.2342e-05 secs\n",
            "\n",
            "\n",
            "Create initial parsimony tree by phylogenetic likelihood library (PLL)... 0.001 seconds\n",
            "Perform fast likelihood tree search using GTR+I+G model...\n",
            "Estimate model parameters (epsilon = 5.000)\n",
            "Perform nearest neighbor interchange...\n",
            "Optimizing NNI: done in 0.0143669 secs using 99.98% CPU\n",
            "Estimate model parameters (epsilon = 1.000)\n",
            "1. Initial log-likelihood: -6139.868\n",
            "Optimal log-likelihood: -6139.613\n",
            "Rate parameters:  A-C: 1.58884  A-G: 5.44799  A-T: 2.42918  C-G: 0.91833  C-T: 3.13300  G-T: 1.00000\n",
            "Base frequencies:  A: 0.209  C: 0.311  G: 0.209  T: 0.270\n",
            "Proportion of invariable sites: 0.221\n",
            "Gamma shape alpha: 0.586\n",
            "Parameters optimization took 1 rounds (0.008 sec)\n",
            "Time for fast ML tree search: 0.051 seconds\n",
            "\n",
            "NOTE: ModelFinder requires 3 MB RAM!\n",
            "ModelFinder will test up to 484 DNA models (sample size: 636) ...\n",
            " No. Model         -LnL         df  AIC          AICc         BIC\n",
            "  1  GTR+F         6724.386     45  13538.772    13545.789    13739.256\n",
            "  2  GTR+F+I       6281.494     46  12654.989    12662.330    12859.928\n",
            "  3  GTR+F+G4      6147.495     46  12386.990    12394.332    12591.929\n",
            "  4  GTR+F+I+G4    6139.547     47  12373.094    12380.768    12582.489\n",
            "  5  GTR+F+R2      6196.785     47  12487.569    12495.243    12696.964\n",
            "  6  GTR+F+R3      6137.050     49  12372.099    12380.461    12590.404\n",
            "  7  GTR+F+R4      6137.119     51  12376.239    12385.321    12603.454\n",
            " 14  GTR+F+I+I+R2  6140.872     48  12377.744    12385.757    12591.593\n",
            " 15  GTR+F+I+I+R3  6137.004     50  12374.007    12382.725    12596.767\n",
            " 25  SYM+G4        6155.703     43  12397.406    12403.798    12588.980\n",
            " 26  SYM+I+G4      6147.038     44  12382.075    12388.776    12578.104\n",
            " 28  SYM+R3        6145.170     46  12382.340    12389.681    12587.279\n",
            " 36  SYM+I+I+R2    6148.911     45  12387.822    12394.839    12588.306\n",
            " 47  TVM+F+G4      6159.347     45  12408.695    12415.711    12609.178\n",
            " 48  TVM+F+I+G4    6150.231     46  12392.463    12399.804    12597.402\n",
            " 50  TVM+F+R3      6147.041     48  12390.081    12398.095    12603.931\n",
            " 58  TVM+F+I+I+R2  6152.284     47  12398.567    12406.241    12607.962\n",
            " 69  TVMe+G4       6155.834     42  12395.669    12401.760    12582.787\n",
            " 70  TVMe+I+G4     6147.099     43  12380.197    12386.589    12571.771\n",
            " 72  TVMe+R3       6145.192     45  12380.384    12387.401    12580.868\n",
            " 80  TVMe+I+I+R2   6149.166     44  12386.331    12393.032    12582.360\n",
            " 91  TIM3+F+G4     6161.162     44  12410.324    12417.024    12606.352\n",
            " 92  TIM3+F+I+G4   6153.745     45  12397.491    12404.508    12597.974\n",
            " 94  TIM3+F+R3     6152.974     47  12399.949    12407.622    12609.343\n",
            "102  TIM3+F+I+I+R2 6153.886     46  12399.773    12407.114    12604.712\n",
            "113  TIM3e+G4      6169.568     41  12421.136    12426.934    12603.800\n",
            "114  TIM3e+I+G4    6161.227     42  12406.455    12412.546    12593.573\n",
            "116  TIM3e+R3      6159.556     44  12407.113    12413.813    12603.141\n",
            "124  TIM3e+I+I+R2  6162.110     43  12410.220    12416.612    12601.794\n",
            "135  TIM2+F+G4     6151.007     44  12390.014    12396.714    12586.042\n",
            "136  TIM2+F+I+G4   6143.387     45  12376.773    12383.790    12577.257\n",
            "138  TIM2+F+R3     6141.066     47  12376.133    12383.806    12585.527\n",
            "146  TIM2+F+I+I+R2 6144.546     46  12381.091    12388.433    12586.030\n",
            "157  TIM2e+G4      6157.097     41  12396.194    12401.992    12578.858\n",
            "158  TIM2e+I+G4    6148.811     42  12381.623    12387.714    12568.741\n",
            "160  TIM2e+R3      6147.128     44  12382.257    12388.957    12578.286\n",
            "168  TIM2e+I+I+R2  6150.646     43  12387.291    12393.683    12578.865\n",
            "179  TIM+F+G4      6163.658     44  12415.316    12422.016    12611.345\n",
            "180  TIM+F+I+G4    6156.435     45  12402.869    12409.886    12603.353\n",
            "182  TIM+F+R3      6155.799     47  12405.598    12413.271    12614.992\n",
            "190  TIM+F+I+I+R2  6156.738     46  12405.477    12412.818    12610.416\n",
            "201  TIMe+G4       6169.977     41  12421.954    12427.752    12604.618\n",
            "202  TIMe+I+G4     6161.790     42  12407.581    12413.672    12594.699\n",
            "204  TIMe+R3       6160.287     44  12408.574    12415.274    12604.602\n",
            "212  TIMe+I+I+R2   6163.048     43  12412.096    12418.488    12603.670\n",
            "223  TPM3u+F+G4    6172.983     43  12431.966    12438.358    12623.539\n",
            "224  TPM3u+F+I+G4  6164.182     44  12416.364    12423.065    12612.393\n",
            "226  TPM3u+F+R3    6162.322     46  12416.644    12423.985    12621.583\n",
            "234  TPM3u+F+I+I+R2 6165.222     45  12420.443    12427.460    12620.927\n",
            "245  TPM3+G4       6169.684     40  12419.368    12424.880    12597.575\n",
            "246  TPM3+I+G4     6161.265     41  12404.530    12410.328    12587.193\n",
            "248  TPM3+R3       6159.555     43  12405.110    12411.502    12596.684\n",
            "256  TPM3+I+I+R2   6162.348     42  12408.696    12414.787    12595.814\n",
            "267  TPM2u+F+G4    6162.420     43  12410.840    12417.232    12602.414\n",
            "268  TPM2u+F+I+G4  6153.825     44  12395.650    12402.350    12591.678\n",
            "270  TPM2u+F+R3    6151.958     46  12395.916    12403.258    12600.855\n",
            "278  TPM2u+F+I+I+R2 6155.787     45  12401.574    12408.591    12602.058\n",
            "289  TPM2+G4       6157.177     40  12394.355    12399.868    12572.563\n",
            "290  TPM2+I+G4     6148.851     41  12379.702    12385.500    12562.366\n",
            "292  TPM2+R3       6147.134     43  12380.268    12386.660    12571.841\n",
            "300  TPM2+I+I+R2   6150.876     42  12385.752    12391.843    12572.870\n",
            "311  K3Pu+F+G4     6175.053     43  12436.106    12442.498    12627.680\n",
            "312  K3Pu+F+I+G4   6166.550     44  12421.100    12427.800    12617.128\n",
            "314  K3Pu+F+R3     6164.939     46  12421.878    12429.220    12626.817\n",
            "322  K3Pu+F+I+I+R2 6167.894     45  12425.788    12432.805    12626.272\n",
            "333  K3P+G4        6170.053     40  12420.107    12425.619    12598.315\n",
            "334  K3P+I+G4      6161.814     41  12405.627    12411.425    12588.290\n",
            "336  K3P+R3        6160.268     43  12406.536    12412.928    12598.109\n",
            "344  K3P+I+I+R2    6163.254     42  12410.509    12416.600    12597.627\n",
            "355  TN+F+G4       6163.961     43  12413.922    12420.314    12605.495\n",
            "356  TN+F+I+G4     6156.801     44  12401.603    12408.303    12597.632\n",
            "358  TN+F+R3       6156.135     46  12404.270    12411.611    12609.209\n",
            "366  TN+F+I+I+R2   6156.966     45  12403.932    12410.948    12604.415\n",
            "377  TNe+G4        6170.212     40  12420.424    12425.937    12598.632\n",
            "378  TNe+I+G4      6162.120     41  12406.240    12412.037    12588.903\n",
            "380  TNe+R3        6160.606     43  12407.213    12413.605    12598.787\n",
            "388  TNe+I+I+R2    6163.234     42  12410.468    12416.559    12597.587\n",
            "399  HKY+F+G4      6175.327     42  12434.655    12440.746    12621.773\n",
            "400  HKY+F+I+G4    6166.916     43  12419.831    12426.223    12611.405\n",
            "402  HKY+F+R3      6165.293     45  12420.586    12427.603    12621.070\n",
            "410  HKY+F+I+I+R2  6168.111     44  12424.221    12430.922    12620.250\n",
            "421  K2P+G4        6170.292     39  12418.585    12423.820    12592.338\n",
            "422  K2P+I+G4      6162.144     40  12404.289    12409.801    12582.497\n",
            "424  K2P+R3        6160.594     42  12405.188    12411.279    12592.307\n",
            "432  K2P+I+I+R2    6163.445     41  12408.891    12414.689    12591.554\n",
            "443  F81+F+G4      6281.119     41  12644.237    12650.035    12826.900\n",
            "444  F81+F+I+G4    6272.465     42  12628.930    12635.021    12816.049\n",
            "446  F81+F+R3      6271.480     44  12630.960    12637.661    12826.989\n",
            "454  F81+F+I+I+R2  6273.343     43  12632.686    12639.078    12824.259\n",
            "465  JC+G4         6278.534     38  12633.067    12638.032    12802.365\n",
            "466  JC+I+G4       6269.637     39  12617.274    12622.509    12791.027\n",
            "468  JC+R3         6268.375     41  12618.750    12624.548    12801.413\n",
            "476  JC+I+I+R2     6271.326     40  12622.652    12628.165    12800.860\n",
            "Akaike Information Criterion:           GTR+F+R3\n",
            "Corrected Akaike Information Criterion: GTR+F+R3\n",
            "Bayesian Information Criterion:         TPM2+I+G4\n",
            "Best-fit model: TPM2+I+G4 chosen according to BIC\n",
            "\n",
            "All model information printed to /tmp/tmpdxraaq6_/q2iqtreeufboot.model.gz\n",
            "CPU time for ModelFinder: 2.435 seconds (0h:0m:2s)\n",
            "Wall-clock time for ModelFinder: 2.447 seconds (0h:0m:2s)\n",
            "Generating 1000 samples for ultrafast bootstrap (seed: 81051)...\n",
            "\n",
            "NOTE: 2 MB RAM (0 GB) is required!\n",
            "Estimate model parameters (epsilon = 0.100)\n",
            "Thoroughly optimizing +I+G parameters from 10 start values...\n",
            "Init pinv, alpha: 0.000, 0.601 / Estimate: 0.000, 0.359 / LogL: -6157.190\n",
            "Init pinv, alpha: 0.049, 0.601 / Estimate: 0.181, 0.505 / LogL: -6149.549\n",
            "Init pinv, alpha: 0.097, 0.601 / Estimate: 0.183, 0.508 / LogL: -6149.506\n",
            "Init pinv, alpha: 0.146, 0.601 / Estimate: 0.182, 0.506 / LogL: -6149.535\n",
            "Init pinv, alpha: 0.195, 0.601 / Estimate: 0.202, 0.536 / LogL: -6149.184\n",
            "Init pinv, alpha: 0.244, 0.601 / Estimate: 0.244, 0.609 / LogL: -6148.842\n",
            "Init pinv, alpha: 0.292, 0.601 / Estimate: 0.278, 0.690 / LogL: -6149.009\n",
            "Init pinv, alpha: 0.341, 0.601 / Estimate: 0.287, 0.715 / LogL: -6149.133\n",
            "Init pinv, alpha: 0.390, 0.601 / Estimate: 0.287, 0.715 / LogL: -6149.129\n",
            "Init pinv, alpha: 0.439, 0.601 / Estimate: 0.287, 0.713 / LogL: -6149.120\n",
            "Optimal pinv,alpha: 0.244, 0.609 / LogL: -6148.842\n",
            "\n",
            "Parameters optimization took 0.314 sec\n",
            "Wrote distance file to... \n",
            "Computing ML distances based on estimated model parameters...\n",
            "Calculating distance matrix: done in 0.00126471 secs using 98.6% CPU\n",
            "Computing ML distances took 0.001305 sec (of wall-clock time) 0.001289 sec (of CPU time)\n",
            "Setting up auxiliary I and S matrices: done in 2.8784e-05 secs using 97.28% CPU\n",
            "Constructing RapidNJ tree: done in 6.3661e-05 secs using 98.96% CPU\n",
            "Computing RapidNJ tree took 0.000116 sec (of wall-clock time) 0.000114 sec (of CPU time)\n",
            "Log-likelihood of RapidNJ tree: -6176.585\n",
            "--------------------------------------------------------------------\n",
            "|             INITIALIZING CANDIDATE TREE SET                      |\n",
            "--------------------------------------------------------------------\n",
            "Generating 98 parsimony trees... 0.080 second\n",
            "Computing log-likelihood of 98 initial trees ... 0.104 seconds\n",
            "Current best score: -6148.842\n",
            "\n",
            "Do NNI search on 20 best initial trees\n",
            "Optimizing NNI: done in 0.00550331 secs using 99.96% CPU\n",
            "Estimate model parameters (epsilon = 0.100)\n",
            "BETTER TREE FOUND at iteration 1: -6148.841\n",
            "Optimizing NNI: done in 0.0108891 secs using 99.92% CPU\n",
            "Optimizing NNI: done in 0.0119528 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0142542 secs using 99.87% CPU\n",
            "Optimizing NNI: done in 0.0188871 secs using 99.95% CPU\n",
            "Optimizing NNI: done in 0.0173716 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0146909 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0165715 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0119009 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0157006 secs using 100% CPU\n",
            "Iteration 10 / LogL: -6148.887 / Time: 0h:0m:0s\n",
            "Optimizing NNI: done in 0.00932806 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0166644 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0133312 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0177261 secs using 77.55% CPU\n",
            "Optimizing NNI: done in 0.0191767 secs using 99.91% CPU\n",
            "Optimizing NNI: done in 0.0148656 secs using 99.96% CPU\n",
            "Optimizing NNI: done in 0.0187742 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0158674 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0196088 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.016603 secs using 99.99% CPU\n",
            "Iteration 20 / LogL: -6159.822 / Time: 0h:0m:0s\n",
            "Finish initializing candidate tree set (5)\n",
            "Current best tree score: -6148.841 / CPU time: 0.496\n",
            "Number of iterations: 20\n",
            "--------------------------------------------------------------------\n",
            "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
            "--------------------------------------------------------------------\n",
            "Optimizing NNI: done in 0.0365184 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0264869 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0220663 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0273475 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0260304 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0142121 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0148946 secs using 73.47% CPU\n",
            "Optimizing NNI: done in 0.0253774 secs using 99.93% CPU\n",
            "Optimizing NNI: done in 0.024644 secs using 83.79% CPU\n",
            "Optimizing NNI: done in 0.0252343 secs using 99.99% CPU\n",
            "Iteration 30 / LogL: -6156.636 / Time: 0h:0m:1s (0h:0m:2s left)\n",
            "Optimizing NNI: done in 0.0149525 secs using 99.94% CPU\n",
            "Optimizing NNI: done in 0.0140506 secs using 99.95% CPU\n",
            "Optimizing NNI: done in 0.0255583 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0198815 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.026609 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0149044 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0195592 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0222227 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.023527 secs using 99.97% CPU\n",
            "Optimizing NNI: done in 0.0225627 secs using 99.99% CPU\n",
            "Iteration 40 / LogL: -6149.444 / Time: 0h:0m:1s (0h:0m:2s left)\n",
            "Optimizing NNI: done in 0.0304683 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.01491 secs using 88.92% CPU\n",
            "Optimizing NNI: done in 0.0140045 secs using 99.81% CPU\n",
            "Optimizing NNI: done in 0.0249953 secs using 99.97% CPU\n",
            "Optimizing NNI: done in 0.0219684 secs using 99.98% CPU\n",
            "Optimizing NNI: done in 0.022268 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0139122 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0183234 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0195846 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0127702 secs using 99.98% CPU\n",
            "Iteration 50 / LogL: -6152.491 / Time: 0h:0m:1s (0h:0m:1s left)\n",
            "Log-likelihood cutoff on original alignment: -6183.229\n",
            "Optimizing NNI: done in 0.0211116 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0141201 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0171993 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0196475 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0236917 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0245196 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0176033 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0190582 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0223688 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0242111 secs using 83.56% CPU\n",
            "Iteration 60 / LogL: -6148.870 / Time: 0h:0m:1s (0h:0m:1s left)\n",
            "Optimizing NNI: done in 0.0222355 secs using 99.93% CPU\n",
            "Optimizing NNI: done in 0.0287253 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0273955 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0262703 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0192674 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0361337 secs using 99.97% CPU\n",
            "Optimizing NNI: done in 0.0251567 secs using 99.97% CPU\n",
            "Optimizing NNI: done in 0.029135 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0257044 secs using 99.98% CPU\n",
            "Optimizing NNI: done in 0.0218237 secs using 81.72% CPU\n",
            "Iteration 70 / LogL: -6151.409 / Time: 0h:0m:2s (0h:0m:0s left)\n",
            "Optimizing NNI: done in 0.0159582 secs using 99.92% CPU\n",
            "Optimizing NNI: done in 0.0171154 secs using 76.54% CPU\n",
            "Optimizing NNI: done in 0.0117378 secs using 99.98% CPU\n",
            "Optimizing NNI: done in 0.0177421 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0250633 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0221831 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0184047 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0213091 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0121906 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0165945 secs using 100% CPU\n",
            "Iteration 80 / LogL: -6149.040 / Time: 0h:0m:2s (0h:0m:0s left)\n",
            "Optimizing NNI: done in 0.018392 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0191247 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0194697 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0281776 secs using 85.81% CPU\n",
            "Optimizing NNI: done in 0.0190184 secs using 99.96% CPU\n",
            "Optimizing NNI: done in 0.0247404 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0200687 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0158598 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.025382 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0272686 secs using 85.35% CPU\n",
            "Iteration 90 / LogL: -6148.991 / Time: 0h:0m:2s (0h:0m:0s left)\n",
            "Optimizing NNI: done in 0.0238 secs using 99.97% CPU\n",
            "Optimizing NNI: done in 0.0145963 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0216946 secs using 99.96% CPU\n",
            "Optimizing NNI: done in 0.0102804 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.0165863 secs using 76% CPU\n",
            "Optimizing NNI: done in 0.0292832 secs using 99.97% CPU\n",
            "Optimizing NNI: done in 0.0199454 secs using 100% CPU\n",
            "Optimizing NNI: done in 0.0248959 secs using 99.94% CPU\n",
            "Optimizing NNI: done in 0.00892054 secs using 99.99% CPU\n",
            "Optimizing NNI: done in 0.019702 secs using 79.74% CPU\n",
            "Iteration 100 / LogL: -6149.388 / Time: 0h:0m:2s (0h:0m:0s left)\n",
            "Log-likelihood cutoff on original alignment: -6183.229\n",
            "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.999\n",
            "Optimizing NNI: done in 0.0135571 secs using 99.98% CPU\n",
            "Optimizing NNI: done in 0.0234263 secs using 99.94% CPU\n",
            "TREE SEARCH COMPLETED AFTER 102 ITERATIONS / Time: 0h:0m:2s\n",
            "\n",
            "--------------------------------------------------------------------\n",
            "|                    FINALIZING TREE SEARCH                        |\n",
            "--------------------------------------------------------------------\n",
            "Performs final model parameters optimization\n",
            "Estimate model parameters (epsilon = 0.010)\n",
            "1. Initial log-likelihood: -6148.841\n",
            "Optimal log-likelihood: -6148.841\n",
            "Rate parameters:  A-C: 1.93775  A-G: 4.12977  A-T: 1.93775  C-G: 1.00000  C-T: 4.12977  G-T: 1.00000\n",
            "Base frequencies:  A: 0.250  C: 0.250  G: 0.250  T: 0.250\n",
            "Proportion of invariable sites: 0.245\n",
            "Gamma shape alpha: 0.611\n",
            "Parameters optimization took 1 rounds (0.003 sec)\n",
            "BEST SCORE FOUND : -6148.841\n",
            "Creating bootstrap support values...\n",
            "Split supports printed to NEXUS file /tmp/tmpdxraaq6_/q2iqtreeufboot.splits.nex\n",
            "Total tree length: 3.890\n",
            "\n",
            "Total number of iterations: 102\n",
            "CPU time used for tree search: 2.445 sec (0h:0m:2s)\n",
            "Wall-clock time used for tree search: 2.496 sec (0h:0m:2s)\n",
            "Total CPU time used: 2.815 sec (0h:0m:2s)\n",
            "Total wall-clock time used: 2.868 sec (0h:0m:2s)\n",
            "\n",
            "Computing bootstrap consensus tree...\n",
            "Reading input file /tmp/tmpdxraaq6_/q2iqtreeufboot.splits.nex...\n",
            "20 taxa and 142 splits.\n",
            "Consensus tree written to /tmp/tmpdxraaq6_/q2iqtreeufboot.contree\n",
            "Reading input trees file /tmp/tmpdxraaq6_/q2iqtreeufboot.contree\n",
            "Log-likelihood of consensus tree: -6148.841\n",
            "\n",
            "Analysis results written to: \n",
            "  IQ-TREE report:                /tmp/tmpdxraaq6_/q2iqtreeufboot.iqtree\n",
            "  Maximum-likelihood tree:       /tmp/tmpdxraaq6_/q2iqtreeufboot.treefile\n",
            "  Likelihood distances:          /tmp/tmpdxraaq6_/q2iqtreeufboot.mldist\n",
            "\n",
            "Ultrafast bootstrap approximation results written to:\n",
            "  Split support values:          /tmp/tmpdxraaq6_/q2iqtreeufboot.splits.nex\n",
            "  Consensus tree:                /tmp/tmpdxraaq6_/q2iqtreeufboot.contree\n",
            "  Screen log file:               /tmp/tmpdxraaq6_/q2iqtreeufboot.log\n",
            "\n",
            "ALISIM COMMAND\n",
            "--------------\n",
            "--alisim simulated_MSA -t /tmp/tmpdxraaq6_/q2iqtreeufboot.treefile -m \"TPM2{1.93775,4.12977}+FQ+I{0.244858}+G4{0.610974}\" --length 636\n",
            "\n",
            "Date and Time: Thu Jun  1 06:56:15 2023\n",
            "Running external command line application. This may print messages to stdout and/or stderr.\n",
            "The command being run is below. This command cannot be manually re-run as it will depend on temporary files that no longer exist.\n",
            "\n",
            "Command: raxmlHPC -f a -m GTRGAMMA -p 3113 -x 4102 -N 100 -s /tmp/qiime2/sc31/data/3267aa82-df09-4459-92db-2277b184680d/data/aligned-dna-sequences.fasta -w /tmp/tmpunmozywe -n q2bootstrap\n",
            "\n",
            "Warning, you specified a working directory via \"-w\"\n",
            "Keep in mind that RAxML only accepts absolute path names, not relative ones!\n",
            "\n",
            "RAxML can't, parse the alignment file as phylip file \n",
            "it will now try to parse it as FASTA file\n",
            "\n",
            "\n",
            "\n",
            "Using BFGS method to optimize GTR rate parameters, to disable this specify \"--no-bfgs\" \n",
            "\n",
            "\n",
            "\n",
            "This is RAxML version 8.2.12 released by Alexandros Stamatakis on May 2018.\n",
            "\n",
            "With greatly appreciated code contributions by:\n",
            "Andre Aberer      (HITS)\n",
            "Simon Berger      (HITS)\n",
            "Alexey Kozlov     (HITS)\n",
            "Kassian Kobert    (HITS)\n",
            "David Dao         (KIT and HITS)\n",
            "Sarah Lutteropp   (KIT and HITS)\n",
            "Nick Pattengale   (Sandia)\n",
            "Wayne Pfeiffer    (SDSC)\n",
            "Akifumi S. Tanabe (NRIFS)\n",
            "Charlie Taylor    (UF)\n",
            "\n",
            "\n",
            "Alignment has 363 distinct alignment patterns\n",
            "\n",
            "Proportion of gaps and completely undetermined characters in this alignment: 5.66%\n",
            "\n",
            "RAxML rapid bootstrapping and subsequent ML search\n",
            "\n",
            "Using 1 distinct models/data partitions with joint branch length optimization\n",
            "\n",
            "\n",
            "\n",
            "Executing 100 rapid bootstrap inferences and thereafter a thorough ML search \n",
            "\n",
            "All free model parameters will be estimated by RAxML\n",
            "GAMMA model of rate heterogeneity, ML estimate of alpha-parameter\n",
            "\n",
            "GAMMA Model parameters will be estimated up to an accuracy of 0.1000000000 Log Likelihood units\n",
            "\n",
            "Partition: 0\n",
            "Alignment Patterns: 363\n",
            "Name: No Name Provided\n",
            "DataType: DNA\n",
            "Substitution Matrix: GTR\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "RAxML was called as follows:\n",
            "\n",
            "raxmlHPC -f a -m GTRGAMMA -p 3113 -x 4102 -N 100 -s /tmp/qiime2/sc31/data/3267aa82-df09-4459-92db-2277b184680d/data/aligned-dna-sequences.fasta -w /tmp/tmpunmozywe -n q2bootstrap \n",
            "\n",
            "\n",
            "\n",
            "Time for BS model parameter optimization 0.165774\n",
            "Bootstrap[0]: Time 0.386363 seconds, bootstrap likelihood -6262.370799, best rearrangement setting 11\n",
            "Bootstrap[1]: Time 0.244831 seconds, bootstrap likelihood -5746.606819, best rearrangement setting 5\n",
            "Bootstrap[2]: Time 0.231330 seconds, bootstrap likelihood -6082.613583, best rearrangement setting 12\n",
            "Bootstrap[3]: Time 0.264247 seconds, bootstrap likelihood -5996.499975, best rearrangement setting 14\n",
            "Bootstrap[4]: Time 0.247325 seconds, bootstrap likelihood -5845.670780, best rearrangement setting 14\n",
            "Bootstrap[5]: Time 0.241218 seconds, bootstrap likelihood -6504.699718, best rearrangement setting 6\n",
            "Bootstrap[6]: Time 0.220707 seconds, bootstrap likelihood -5999.484894, best rearrangement setting 6\n",
            "Bootstrap[7]: Time 0.250754 seconds, bootstrap likelihood -5885.940458, best rearrangement setting 5\n",
            "Bootstrap[8]: Time 0.235261 seconds, bootstrap likelihood -6538.668064, best rearrangement setting 13\n",
            "Bootstrap[9]: Time 0.224314 seconds, bootstrap likelihood -6374.941742, best rearrangement setting 12\n",
            "Bootstrap[10]: Time 0.236360 seconds, bootstrap likelihood -6354.004146, best rearrangement setting 7\n",
            "Bootstrap[11]: Time 0.229583 seconds, bootstrap likelihood -6096.729866, best rearrangement setting 6\n",
            "Bootstrap[12]: Time 0.250132 seconds, bootstrap likelihood -6452.211078, best rearrangement setting 5\n",
            "Bootstrap[13]: Time 0.230909 seconds, bootstrap likelihood -5990.158906, best rearrangement setting 8\n",
            "Bootstrap[14]: Time 0.259740 seconds, bootstrap likelihood -6085.661018, best rearrangement setting 11\n",
            "Bootstrap[15]: Time 0.245722 seconds, bootstrap likelihood -6142.200793, best rearrangement setting 15\n",
            "Bootstrap[16]: Time 0.209653 seconds, bootstrap likelihood -5677.283913, best rearrangement setting 7\n",
            "Bootstrap[17]: Time 0.195957 seconds, bootstrap likelihood -6199.364055, best rearrangement setting 8\n",
            "Bootstrap[18]: Time 0.267532 seconds, bootstrap likelihood -5982.010175, best rearrangement setting 13\n",
            "Bootstrap[19]: Time 0.290879 seconds, bootstrap likelihood -6392.091333, best rearrangement setting 11\n",
            "Bootstrap[20]: Time 0.245480 seconds, bootstrap likelihood -6236.329498, best rearrangement setting 14\n",
            "Bootstrap[21]: Time 0.208535 seconds, bootstrap likelihood -6186.778007, best rearrangement setting 12\n",
            "Bootstrap[22]: Time 0.200367 seconds, bootstrap likelihood -6188.845261, best rearrangement setting 8\n",
            "Bootstrap[23]: Time 0.248340 seconds, bootstrap likelihood -5867.694111, best rearrangement setting 6\n",
            "Bootstrap[24]: Time 0.263888 seconds, bootstrap likelihood -6027.722834, best rearrangement setting 11\n",
            "Bootstrap[25]: Time 0.230818 seconds, bootstrap likelihood -6055.229067, best rearrangement setting 8\n",
            "Bootstrap[26]: Time 0.223034 seconds, bootstrap likelihood -6117.893231, best rearrangement setting 9\n",
            "Bootstrap[27]: Time 0.244482 seconds, bootstrap likelihood -6531.801350, best rearrangement setting 10\n",
            "Bootstrap[28]: Time 0.219435 seconds, bootstrap likelihood -5900.662713, best rearrangement setting 5\n",
            "Bootstrap[29]: Time 0.250707 seconds, bootstrap likelihood -5765.442107, best rearrangement setting 7\n",
            "Bootstrap[30]: Time 0.244909 seconds, bootstrap likelihood -5709.072639, best rearrangement setting 8\n",
            "Bootstrap[31]: Time 0.243512 seconds, bootstrap likelihood -5785.012330, best rearrangement setting 8\n",
            "Bootstrap[32]: Time 0.227345 seconds, bootstrap likelihood -5916.812112, best rearrangement setting 10\n",
            "Bootstrap[33]: Time 0.221276 seconds, bootstrap likelihood -6013.768758, best rearrangement setting 9\n",
            "Bootstrap[34]: Time 0.243802 seconds, bootstrap likelihood -6439.832967, best rearrangement setting 5\n",
            "Bootstrap[35]: Time 0.239510 seconds, bootstrap likelihood -6147.140024, best rearrangement setting 6\n",
            "Bootstrap[36]: Time 0.217639 seconds, bootstrap likelihood -6149.920373, best rearrangement setting 12\n",
            "Bootstrap[37]: Time 0.285311 seconds, bootstrap likelihood -5925.049726, best rearrangement setting 14\n",
            "Bootstrap[38]: Time 0.238368 seconds, bootstrap likelihood -6009.094037, best rearrangement setting 9\n",
            "Bootstrap[39]: Time 0.218553 seconds, bootstrap likelihood -6283.806674, best rearrangement setting 5\n",
            "Bootstrap[40]: Time 0.217003 seconds, bootstrap likelihood -6175.620662, best rearrangement setting 12\n",
            "Bootstrap[41]: Time 0.250521 seconds, bootstrap likelihood -5967.635618, best rearrangement setting 7\n",
            "Bootstrap[42]: Time 0.235759 seconds, bootstrap likelihood -5864.907152, best rearrangement setting 9\n",
            "Bootstrap[43]: Time 0.206238 seconds, bootstrap likelihood -6078.196943, best rearrangement setting 5\n",
            "Bootstrap[44]: Time 0.258009 seconds, bootstrap likelihood -6098.514341, best rearrangement setting 8\n",
            "Bootstrap[45]: Time 0.243495 seconds, bootstrap likelihood -5929.551205, best rearrangement setting 8\n",
            "Bootstrap[46]: Time 0.228470 seconds, bootstrap likelihood -6073.735643, best rearrangement setting 9\n",
            "Bootstrap[47]: Time 0.259768 seconds, bootstrap likelihood -6034.838309, best rearrangement setting 10\n",
            "Bootstrap[48]: Time 0.217688 seconds, bootstrap likelihood -6199.294570, best rearrangement setting 14\n",
            "Bootstrap[49]: Time 0.252082 seconds, bootstrap likelihood -6712.411503, best rearrangement setting 10\n",
            "Bootstrap[50]: Time 0.218766 seconds, bootstrap likelihood -6146.302168, best rearrangement setting 14\n",
            "Bootstrap[51]: Time 0.238151 seconds, bootstrap likelihood -5869.065262, best rearrangement setting 5\n",
            "Bootstrap[52]: Time 0.244212 seconds, bootstrap likelihood -6142.675331, best rearrangement setting 14\n",
            "Bootstrap[53]: Time 0.219783 seconds, bootstrap likelihood -6125.879616, best rearrangement setting 8\n",
            "Bootstrap[54]: Time 0.237001 seconds, bootstrap likelihood -6313.061074, best rearrangement setting 14\n",
            "Bootstrap[55]: Time 0.223370 seconds, bootstrap likelihood -6156.928413, best rearrangement setting 10\n",
            "Bootstrap[56]: Time 0.245446 seconds, bootstrap likelihood -5909.452587, best rearrangement setting 6\n",
            "Bootstrap[57]: Time 0.243680 seconds, bootstrap likelihood -6381.267072, best rearrangement setting 12\n",
            "Bootstrap[58]: Time 0.268153 seconds, bootstrap likelihood -6160.753964, best rearrangement setting 7\n",
            "Bootstrap[59]: Time 0.274944 seconds, bootstrap likelihood -6261.173148, best rearrangement setting 9\n",
            "Bootstrap[60]: Time 0.296416 seconds, bootstrap likelihood -5990.043965, best rearrangement setting 15\n",
            "Bootstrap[61]: Time 0.245499 seconds, bootstrap likelihood -6034.387352, best rearrangement setting 13\n",
            "Bootstrap[62]: Time 0.254828 seconds, bootstrap likelihood -5859.812133, best rearrangement setting 7\n",
            "Bootstrap[63]: Time 0.206736 seconds, bootstrap likelihood -5993.056938, best rearrangement setting 7\n",
            "Bootstrap[64]: Time 0.233563 seconds, bootstrap likelihood -5968.571356, best rearrangement setting 10\n",
            "Bootstrap[65]: Time 0.225809 seconds, bootstrap likelihood -5854.910900, best rearrangement setting 6\n",
            "Bootstrap[66]: Time 0.210940 seconds, bootstrap likelihood -5992.255451, best rearrangement setting 15\n",
            "Bootstrap[67]: Time 0.244260 seconds, bootstrap likelihood -5832.105908, best rearrangement setting 12\n",
            "Bootstrap[68]: Time 0.267254 seconds, bootstrap likelihood -6058.887955, best rearrangement setting 13\n",
            "Bootstrap[69]: Time 0.230384 seconds, bootstrap likelihood -6151.044163, best rearrangement setting 11\n",
            "Bootstrap[70]: Time 0.227021 seconds, bootstrap likelihood -6152.583159, best rearrangement setting 12\n",
            "Bootstrap[71]: Time 0.221271 seconds, bootstrap likelihood -5880.731835, best rearrangement setting 13\n",
            "Bootstrap[72]: Time 0.267531 seconds, bootstrap likelihood -5803.590618, best rearrangement setting 7\n",
            "Bootstrap[73]: Time 0.239042 seconds, bootstrap likelihood -5947.341521, best rearrangement setting 6\n",
            "Bootstrap[74]: Time 0.268623 seconds, bootstrap likelihood -6308.701843, best rearrangement setting 11\n",
            "Bootstrap[75]: Time 0.259154 seconds, bootstrap likelihood -6307.177952, best rearrangement setting 5\n",
            "Bootstrap[76]: Time 0.245766 seconds, bootstrap likelihood -6084.447988, best rearrangement setting 9\n",
            "Bootstrap[77]: Time 0.240174 seconds, bootstrap likelihood -6133.442893, best rearrangement setting 5\n",
            "Bootstrap[78]: Time 0.238240 seconds, bootstrap likelihood -6193.187369, best rearrangement setting 10\n",
            "Bootstrap[79]: Time 0.239783 seconds, bootstrap likelihood -6292.560571, best rearrangement setting 9\n",
            "Bootstrap[80]: Time 0.231352 seconds, bootstrap likelihood -6178.007778, best rearrangement setting 5\n",
            "Bootstrap[81]: Time 0.270198 seconds, bootstrap likelihood -6381.012968, best rearrangement setting 5\n",
            "Bootstrap[82]: Time 0.271499 seconds, bootstrap likelihood -6779.493985, best rearrangement setting 5\n",
            "Bootstrap[83]: Time 0.230652 seconds, bootstrap likelihood -5818.456620, best rearrangement setting 15\n",
            "Bootstrap[84]: Time 0.232615 seconds, bootstrap likelihood -6267.319336, best rearrangement setting 13\n",
            "Bootstrap[85]: Time 0.249263 seconds, bootstrap likelihood -6345.600860, best rearrangement setting 14\n",
            "Bootstrap[86]: Time 0.230499 seconds, bootstrap likelihood -6365.005184, best rearrangement setting 10\n",
            "Bootstrap[87]: Time 0.228972 seconds, bootstrap likelihood -6312.216260, best rearrangement setting 13\n",
            "Bootstrap[88]: Time 0.196511 seconds, bootstrap likelihood -5953.712308, best rearrangement setting 12\n",
            "Bootstrap[89]: Time 0.224590 seconds, bootstrap likelihood -5850.786668, best rearrangement setting 12\n",
            "Bootstrap[90]: Time 0.263728 seconds, bootstrap likelihood -6033.913828, best rearrangement setting 12\n",
            "Bootstrap[91]: Time 0.253600 seconds, bootstrap likelihood -6170.696167, best rearrangement setting 12\n",
            "Bootstrap[92]: Time 0.230539 seconds, bootstrap likelihood -5912.476423, best rearrangement setting 15\n",
            "Bootstrap[93]: Time 0.220643 seconds, bootstrap likelihood -6001.321707, best rearrangement setting 15\n",
            "Bootstrap[94]: Time 0.237041 seconds, bootstrap likelihood -6083.858894, best rearrangement setting 10\n",
            "Bootstrap[95]: Time 0.253100 seconds, bootstrap likelihood -5943.032506, best rearrangement setting 9\n",
            "Bootstrap[96]: Time 0.219115 seconds, bootstrap likelihood -5925.740999, best rearrangement setting 14\n",
            "Bootstrap[97]: Time 0.244077 seconds, bootstrap likelihood -5884.270150, best rearrangement setting 7\n",
            "Bootstrap[98]: Time 0.225249 seconds, bootstrap likelihood -6132.343900, best rearrangement setting 9\n",
            "Bootstrap[99]: Time 0.216428 seconds, bootstrap likelihood -6031.091860, best rearrangement setting 11\n",
            "\n",
            "\n",
            "Overall Time for 100 Rapid Bootstraps 24.058421 seconds\n",
            "Average Time per Rapid Bootstrap 0.240584 seconds\n",
            "\n",
            "Starting ML Search ...\n",
            "\n",
            "Fast ML optimization finished\n",
            "\n",
            "Fast ML search Time: 23.386898 seconds\n",
            "\n",
            "Slow ML Search 0 Likelihood: -6147.484859\n",
            "Slow ML Search 1 Likelihood: -6147.484859\n",
            "Slow ML Search 2 Likelihood: -6147.484859\n",
            "Slow ML Search 3 Likelihood: -6147.484859\n",
            "Slow ML Search 4 Likelihood: -6147.484859\n",
            "Slow ML Search 5 Likelihood: -6147.484859\n",
            "Slow ML Search 6 Likelihood: -6147.484859\n",
            "Slow ML Search 7 Likelihood: -6147.484859\n",
            "Slow ML Search 8 Likelihood: -6147.484859\n",
            "Slow ML Search 9 Likelihood: -6147.484859\n",
            "Slow ML optimization finished\n",
            "\n",
            "Slow ML search Time: 6.002274 seconds\n",
            "Thorough ML search Time: 0.707324 seconds\n",
            "\n",
            "Final ML Optimization Likelihood: -6147.483800\n",
            "\n",
            "Model Information:\n",
            "\n",
            "Model Parameters of Partition 0, Name: No Name Provided, Type of Data: DNA\n",
            "alpha: 0.364997\n",
            "Tree-Length: 4.066331\n",
            "rate A <-> C: 1.609666\n",
            "rate A <-> G: 5.605683\n",
            "rate A <-> T: 2.427847\n",
            "rate C <-> G: 0.935412\n",
            "rate C <-> T: 3.111797\n",
            "rate G <-> T: 1.000000\n",
            "\n",
            "freq pi(A): 0.209333\n",
            "freq pi(C): 0.311333\n",
            "freq pi(G): 0.209250\n",
            "freq pi(T): 0.270083\n",
            "\n",
            "\n",
            "ML search took 30.096669 secs or 0.008360 hours\n",
            "\n",
            "Combined Bootstrap and ML search took 54.155097 secs or 0.015043 hours\n",
            "\n",
            "Drawing Bootstrap Support Values on best-scoring ML tree ...\n",
            "\n",
            "\n",
            "\n",
            "Found 1 tree in File /tmp/tmpunmozywe/RAxML_bestTree.q2bootstrap\n",
            "\n",
            "\n",
            "\n",
            "Found 1 tree in File /tmp/tmpunmozywe/RAxML_bestTree.q2bootstrap\n",
            "\n",
            "Program execution info written to /tmp/tmpunmozywe/RAxML_info.q2bootstrap\n",
            "All 100 bootstrapped trees written to: /tmp/tmpunmozywe/RAxML_bootstrap.q2bootstrap\n",
            "\n",
            "Best-scoring ML tree written to: /tmp/tmpunmozywe/RAxML_bestTree.q2bootstrap\n",
            "\n",
            "Best-scoring ML tree with support values written to: /tmp/tmpunmozywe/RAxML_bipartitions.q2bootstrap\n",
            "\n",
            "Best-scoring ML tree with support values as branch labels written to: /tmp/tmpunmozywe/RAxML_bipartitionsBranchLabels.q2bootstrap\n",
            "\n",
            "Overall execution time for full ML analysis: 54.155966 secs or 0.015043 hours or 0.000627 days\n",
            "\n",
            "Running external command line application. This may print messages to stdout and/or stderr.\n",
            "The command being run is below. This command cannot be manually re-run as it will depend on temporary files that no longer exist.\n",
            "\n",
            "Command: FastTree -quote -nt /tmp/qiime2/sc31/data/3267aa82-df09-4459-92db-2277b184680d/data/aligned-dna-sequences.fasta\n",
            "\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "FastTree Version 2.1.11 Double precision (No SSE3)\n",
            "Alignment: /tmp/qiime2/sc31/data/3267aa82-df09-4459-92db-2277b184680d/data/aligned-dna-sequences.fasta\n",
            "Nucleotide distances: Jukes-Cantor Joins: balanced Support: SH-like 1000\n",
            "Search: Normal +NNI +SPR (2 rounds range 10) +ML-NNI opt-each=1\n",
            "TopHits: 1.00*sqrtN close=default refresh=0.80\n",
            "ML Model: Jukes-Cantor, CAT approximation with 20 rate categories\n",
            "Initial topology in 0.00 seconds\n",
            "Refining topology: 17 rounds ME-NNIs, 2 rounds ME-SPRs, 9 rounds ML-NNIs\n",
            "Total branch-length 2.047 after 0.04 sec\n",
            "ML-NNI round 1: LogLk = -6854.969 NNIs 5 max delta 5.69 Time 0.05\n",
            "Switched to using 20 rate categories (CAT approximation)\n",
            "Rate categories were divided by 0.823 so that average rate = 1.0\n",
            "CAT-based log-likelihoods may not be comparable across runs\n",
            "Use -gamma for approximate but comparable Gamma(20) log-likelihoods\n",
            "ML-NNI round 2: LogLk = -6015.470 NNIs 1 max delta 6.30 Time 0.07\n",
            "ML-NNI round 3: LogLk = -6015.351 NNIs 0 max delta 0.00 Time 0.07\n",
            "Turning off heuristics for final round of ML NNIs (converged)\n",
            "ML-NNI round 4: LogLk = -6013.551 NNIs 0 max delta 0.00 Time 0.09 (final)\n",
            "Optimize all lengths: LogLk = -6013.523 Time 0.09\n",
            "Total time: 0.11 seconds Unique: 20/20 Bad splits: 0/17\n"
          ]
        }
      ],
      "source": [
        "# First, generate a multiple sequence alignment with mafft\n",
        "alignedotus = mafft(otuSeqs)\n",
        "iqtreeUR = pm.iqtree_ultrafast_bootstrap(alignedotus.alignment)\n",
        "raxmlUR = pm.raxml_rapid_bootstrap(alignedotus.alignment)\n",
        "fasttreeUR = pm.fasttree(alignedotus.alignment)\n",
        "iqtreeR = pm.midpoint_root(iqtreeUR.tree)\n",
        "raxmlR = pm.midpoint_root(raxmlUR.tree)\n",
        "fasttreeR = pm.midpoint_root(fasttreeUR.tree)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 77,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'artifacts/fasttree.qza'"
            ]
          },
          "execution_count": 77,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "iqtreeUR.tree.save('artifacts/iqtreeUR.qza')\n",
        "raxmlUR.tree.save('artifacts/raxmlUR.qza')\n",
        "fasttreeUR.tree.save('artifacts/fasttreeUR.qza')\n",
        "iqtreeR.rooted_tree.save('artifacts/iqtreeR.qza')\n",
        "raxmlR.rooted_tree.save('artifacts/raxmlR.qza')\n",
        "fasttreeR.rooted_tree.save('artifacts/fasttree.qza')"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Data sources\n",
        "\n",
        "- https://www.ebi.ac.uk/ena/browser/view/PRJEB8518\n",
        "- https://docs.qiime2.org/2023.5/data-resources/ High-quality reference OTUs\n",
        "- The `see` and `v2frame` functions were obtained from [this](https://forum.qiime2.org/t/how-to-capture-a-value-from-a-summary-and-pipe-it/19783/5) link via user thermokarst\n"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Load previous artifacts"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 78,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Load previous artifacts\n",
        "denoised_Seqs = Artifact.load('artifacts/denoised_seqs.qza')\n",
        "denoised_Ftable = Artifact.load('artifacts/denoised.qza')\n",
        "otuFreqs = Artifact.load('artifacts/otuFreqs.qza')\n",
        "otuSeqs = Artifact.load('artifacts/otuSeqs.qza')\n",
        "ref_ids = Artifact.load('../data/artifacts/AllIDs.qza')\n",
        "ref_seqs = Artifact.load('../data/artifacts/AllSeqs.qza')\n",
        "blast = Artifact.load('artifacts/blast_seqs.qza')\n",
        "blast_top_hits = Artifact.load('artifacts/blast_hits.qza')\n",
        "vsearch_seqs = Artifact.load('artifacts/vsearch_seqs.qza')\n",
        "vsearch_top_hits = Artifact.load('artifacts/vsearch_hits.qza')\n",
        "iqtreeUR = Artifact.load('artifacts/iqtreeUR.qza')\n",
        "raxmlUR = Artifact.load('artifacts/raxmlUR.qza')\n",
        "fasttreeUR = Artifact.load('artifacts/fasttreeUR.qza')\n",
        "iqtreeR = Artifact.load('artifacts/iqtreeR.qza')\n",
        "raxmlR = Artifact.load('artifacts/raxmlR.qza')\n",
        "fasttreeR = Artifact.load('artifacts/fasttreeR.qza')"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8.15"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 4
}
