{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### Jellyfish microbiome analysis\n",
        "\n",
        "#### Sample background"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Essential objects\n",
        "from qiime2 import Metadata\n",
        "import pandas as pd\n",
        "from qiime2 import Artifact, sdk\n",
        "from qiime2.plugins.dada2.methods import denoise_pyro # The samples were obtained through pyrosequencing\n",
        "import qiime2.plugins.metadata.actions as metadata_actions\n",
        "import qiime2.plugins.feature_classifier.actions as feature_classifier_actions\n",
        "from qiime2.plugins.feature_table.visualizers import tabulate_seqs\n",
        "import qiime2.plugins.feature_table.actions as fta\n",
        "import matplotlib.pyplot as plt\n",
        "import matplotlib \n",
        "import seaborn as sns\n",
        "from Bio import SeqIO\n",
        "\n",
        "matplotlib.use('module://ipykernel.pylab.backend_inline')\n",
        "pm = sdk.PluginManager()\n",
        "def see(artifact):\n",
        "    from_format = artifact.format\n",
        "    if issubclass(from_format, sdk.plugin_manager.SingleFileDirectoryFormatBase):\n",
        "        from_format = artifact.format.file.format\n",
        "    return set(pm.transformers[from_format].keys())\n",
        "import os\n",
        "import pandas\n",
        "import qiime2\n",
        "import tempfile\n",
        "\n",
        "def v2frame(viz_fp: str) -> list:\n",
        "    '''viz_fp is a path to the qiime2 visualization object'''\n",
        "    viz = qiime2.Visualization.load(viz_fp)\n",
        "    with tempfile.TemporaryDirectory() as tmpdir:\n",
        "        viz.export_data(tmpdir)\n",
        "        fp = os.path.join(tmpdir, 'quality-plot.html')\n",
        "        ov = os.path.join(tmpdir, 'overview.html')\n",
        "        dfs = pandas.read_html(fp, index_col=0)\n",
        "        df2s = pandas.read_html(ov, index_col=0)\n",
        "    return dfs + df2s"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### Importing the data\n",
        "The fastq files contain single-end reads, sequenced with the 454 GS FLX+ via a pyrosequencing approach\n",
        "\n",
        "```{bash}\n",
        "qiime tools import \\\n",
        "    --type 'SampleData[SequencesWithQuality]' \\\n",
        "    --input-path data.tsv \\\n",
        "    --output-path jelly.qza \\\n",
        "    --input-format SingleEndFastqManifestPhred33V2\n",
        "```"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [],
      "source": [
        "raw_data = Artifact.load('artifacts/jelly.qza')"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### Raw data exploration\n",
        "The initial data exploration was performed with fastqc. There is a high variation in read length, though the majority are in the 450 - 500 bp range. As expected for short-read sequencing, read quality drops at higher read lengths, necessitating trimming. Surprisingly, the sample has little adapter content.\n",
        "\n",
        "#### Blasting overrepresenting sequences\n",
        "To get a rough idea of what microorganisms were represented most in the sample, I used data from `fastqc`'s overrepresented sequences module.\n",
        "The sequences listed in the module were combined, clustered together to remove redundant reads with `cd-hit`, then blasted against the custom 16s rRNA database.\n",
        "- `cd-hit` grouped 238 overrepresented sequences from the 7 samples into only 17 clusters\n",
        "- The following describes the best 3 hits for the top 3 most overrepresented sequences\n",
        "    - 1. Entoplasma, Mesoplasma, Lebetimonas\n",
        "    - 2. *Ferruginivarius sediminum*, Azospirillum, Desulfosporosinus\n",
        "    - 3. Flavobacteria, Dokdonia, Joostella\n",
        "\n"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### Quality control and clustering\n",
        "The dada2 plugin denoises sequences and clusters them into representative sequences"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
            "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
            "\n",
            "Command: run_dada.R --input_directory /tmp/qiime2/sc31/data/54baf4e3-b7ca-4f9c-a76c-8897cbea851c/data --output_path /tmp/tmplrl4aal9/output.tsv.biom --output_track /tmp/tmplrl4aal9/track.tsv --filtered_directory /tmp/tmplrl4aal9 --truncation_length 600 --trim_left 0 --max_expected_errors 2.0 --truncation_quality_score 2 --max_length Inf --pooling_method independent --chimera_method consensus --min_parental_fold 1.0 --allow_one_off False --num_threads 1 --learn_min_reads 250000 --homopolymer_gap_penalty 1 --band_size 32\n",
            "\n",
            "R version 4.2.2 (2022-10-31) \n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Loading required package: Rcpp\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "DADA2: 1.26.0 / Rcpp: 1.0.10 / RcppParallel: 5.1.6 \n",
            "2) Filtering .......\n",
            "3) Learning Error Rates\n",
            "11095200 total bases in 18492 reads from 7 samples will be used for learning the error rates.\n",
            "4) Denoise samples \n",
            ".......\n",
            "5) Remove chimeras (method = consensus)\n",
            "6) Report read numbers through the pipeline\n",
            "7) Write output\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "'vis/denoised.qzv'"
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "denoised_Ftable, denoised_Seqs, denoised_stats = denoise_pyro(raw_data, trunc_len=600) # Will begin filtering at 600 bp\n",
        "denoised_Ftable.save('artifacts/denoised.qza')\n",
        "denoised_Seqs.save('artifacts/denoised_seqs.qza')\n",
        "stats_viz, = metadata_actions.tabulate(input=denoised_stats.view(Metadata))\n",
        "stats_viz.save('vis/denoised.qzv')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "FeatureData[Sequence]\n",
            "FeatureTable[Frequency]\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "'vis/sequences.qzv'"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Importing previous\n",
        "denoised_Seqs = Artifact.load('artifacts/denoised_seqs.qza')\n",
        "print(denoised_Seqs.type)\n",
        "denoised_Ftable = Artifact.load('artifacts/denoised.qza')\n",
        "print(denoised_Ftable.type)\n",
        "denoised_FtableDF = denoised_Ftable.view(pd.DataFrame).T\n",
        "denoised_FtableDF.index.names = ['Feature ID']\n",
        "denoised_FtableMD = denoised_Ftable.view(Metadata)\n",
        "seqs_viz, = fta.tabulate_seqs(data=denoised_Seqs) # Tabulate sequences to \n",
        "denoised_SeqsDF = denoised_Seqs.view(Metadata).to_dataframe()\n",
        "seqs_viz.save('vis/sequences.qzv')"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### Pairing sequences with their frequencies in Pandas"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>sample-1</th>\n",
              "      <th>sample-2</th>\n",
              "      <th>sample-3</th>\n",
              "      <th>sample-4</th>\n",
              "      <th>sample-5</th>\n",
              "      <th>sample-6</th>\n",
              "      <th>sample-7</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Feature ID</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>92.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>5247.0</td>\n",
              "      <td>1983.0</td>\n",
              "      <td>CACTCTTGCGAGCATACTACTCAGGCGGAGTACTTAACGCGTTAGC...</td>\n",
              "      <td>8331adf13c184313c464be9c7b21097c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>4268.0</td>\n",
              "      <td>2209.0</td>\n",
              "      <td>325.0</td>\n",
              "      <td>392.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>CACTCTTGCGAGCATACTACTCAGGCGGAGTACTTAACGCGTTAGC...</td>\n",
              "      <td>b37b36ef2cd58aa19929a8ae4f8f4c21</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>671.0</td>\n",
              "      <td>277.0</td>\n",
              "      <td>97.0</td>\n",
              "      <td>30.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>387.0</td>\n",
              "      <td>42.0</td>\n",
              "      <td>CATTCTTGCGAACGTACTCCCCAGGTGGGATACTTATCACTTTCGC...</td>\n",
              "      <td>b5f4256d637420120c2e71bc85666bfd</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>94.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>298.0</td>\n",
              "      <td>284.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>6.0</td>\n",
              "      <td>TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...</td>\n",
              "      <td>00aabc54895675bbc65b0c1fdaf1fd07</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>681.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>CACTCTTGCGAGCATACTACTCAGGCGGAGTACTTAACGCGTTAGC...</td>\n",
              "      <td>3c5ee7863e02dd5604419d803a014014</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>235.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>121.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>CACACTTGCGTGCGTACTCCCCAGGCGGAACACTTAACGCGTTGGC...</td>\n",
              "      <td>fe008f8f1f5a1776177e14bae97e10e2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>188.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAATCTTGCGACCGTACTCCCCAGGCGGAATGCTTAATCCGTTAGG...</td>\n",
              "      <td>3a4c4c7d2b8e5a489eb2d1fddc95e025</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>163.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...</td>\n",
              "      <td>0c2db32fc52f027a04e433a93c0669df</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>76.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...</td>\n",
              "      <td>5f460fa8105db752feb6166a52b135d2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>45.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAGTCTTGCGACCGTAGTCCCCAGGCGGAGTGCTTAACGCGTTAGC...</td>\n",
              "      <td>6d4121b0c9e7eda284b4ab5bf571dc86</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>16.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>CACACTTGCGTGCGTACTCCCCAGGCGGAACACTTAACGCGTTGGC...</td>\n",
              "      <td>6e0740a41b665e09ed40342b621fb3b0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>13.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAGTCTTGCGACCGTACTCCCCAGGCGGAGAACTTAACGCGTTAGC...</td>\n",
              "      <td>c6eddca0dc7ee0b43403e873ba6c8cfc</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>11.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAATCTTGCGACCGTACTCCCCAGGCGGTCTACTTAACGCGTTAGC...</td>\n",
              "      <td>d4c1d85ff4ca5efdc346801641b2d9a2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>10.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>CAGCCTTGCGACCATACTCCCCAGGCGGAACACTTAACGCTTTCGC...</td>\n",
              "      <td>c00ee893109657877cc78cca52003ca9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>9.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAATCTTGCGACCGTACTCCCCAGGCGGTTCATTTAATGCGTTAGC...</td>\n",
              "      <td>d6f1d18cadc754defc476dd14e4f4c58</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>8.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>CATTCTTGCGAACGTACTCCCCAGGTGGGATACTTATCACTTTCGC...</td>\n",
              "      <td>ac30e74e21dd088eabfea4d8738c6279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>7.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAACCTTGCGGCCGTACTCCCCAGGCGGTCAACTTATCGCGTTAGC...</td>\n",
              "      <td>4d0e9bcbdf46ba60e8c807aca0550f26</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>5.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAGCCTTGCGGCCGTACTCCCCAGGCGGTACACTTATCGCGTTAGC...</td>\n",
              "      <td>3046631d2ae55b33a73e710aecb4db9e</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>0.0</td>\n",
              "      <td>5.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAATCTTGCGACCGTACTCCCCAGGCGGAGTGCTTAATGCGTTAGC...</td>\n",
              "      <td>6ae71cd37d021784c75d9c4ca7de333d</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>4.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>CAACCTTGCGGCCGTACTCCCCAGGCGGAGTGCTTAATGCGTTAGC...</td>\n",
              "      <td>c27592c0acdd2708a5b18cf790d0d1e1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>4.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAGCCTTGCGGCCGTACTCCCCAGGCGGATAACTTAACGCGTTAGC...</td>\n",
              "      <td>a6c50d7fef8a40824552274d0e96e3c4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>4.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAGCCTTGCGGCCGTACTCCCCAGGCGGACAACTTAAAACGTTAGC...</td>\n",
              "      <td>f67e6cc781a78a69a56d9d441411694b</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>3.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>CATTCTTGCGAACGTACTCCCCAGGTGGGATACTTATCACTTTCGC...</td>\n",
              "      <td>c92bce272a01f65e7105823ab7957ae8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>3.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAATCTTGCGACCGTACTCCCCAGGCGGTCTACTTATCGCGTTAGC...</td>\n",
              "      <td>c74f9ca1eae5081f61f84934c8e22a79</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>3.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...</td>\n",
              "      <td>b770c9ce6a8ca4de3d7085854747551c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>2.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>TAGCCTTGCGGCCGTACTCCCCAGGCGGAGAACTTATCGCGTTAGC...</td>\n",
              "      <td>f1a3e9623a42e1ee02b2317bb9a02a25</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "    sample-1  sample-2  sample-3  sample-4  sample-5  sample-6  sample-7  \\\n",
              "0        0.0       0.0      92.0       0.0       0.0    5247.0    1983.0   \n",
              "1     4268.0    2209.0     325.0     392.0       0.0       0.0       0.0   \n",
              "2      671.0     277.0      97.0      30.0       0.0     387.0      42.0   \n",
              "3       94.0       0.0     298.0     284.0       0.0       0.0       6.0   \n",
              "4        0.0       0.0       0.0       0.0     681.0       0.0       0.0   \n",
              "5      235.0       0.0       0.0       0.0       0.0     121.0       0.0   \n",
              "6      188.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
              "7        0.0       0.0       0.0       0.0       0.0     163.0       0.0   \n",
              "8        0.0       0.0       0.0       0.0      76.0       0.0       0.0   \n",
              "9        0.0       0.0       0.0       0.0       0.0      45.0       0.0   \n",
              "10       0.0       0.0      16.0       0.0       0.0       0.0       0.0   \n",
              "11       0.0       0.0       0.0       0.0       0.0      13.0       0.0   \n",
              "12      11.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
              "13      10.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
              "14       0.0       0.0       0.0       0.0       0.0       9.0       0.0   \n",
              "15       0.0       0.0       0.0       0.0       8.0       0.0       0.0   \n",
              "16       0.0       0.0       0.0       0.0       0.0       7.0       0.0   \n",
              "17       5.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
              "18       0.0       5.0       0.0       0.0       0.0       0.0       0.0   \n",
              "19       4.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
              "20       4.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
              "21       0.0       0.0       4.0       0.0       0.0       0.0       0.0   \n",
              "22       3.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
              "23       3.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
              "24       0.0       0.0       0.0       3.0       0.0       0.0       0.0   \n",
              "25       0.0       0.0       0.0       0.0       0.0       2.0       0.0   \n",
              "\n",
              "                                             Sequence  \\\n",
              "0   CACTCTTGCGAGCATACTACTCAGGCGGAGTACTTAACGCGTTAGC...   \n",
              "1   CACTCTTGCGAGCATACTACTCAGGCGGAGTACTTAACGCGTTAGC...   \n",
              "2   CATTCTTGCGAACGTACTCCCCAGGTGGGATACTTATCACTTTCGC...   \n",
              "3   TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...   \n",
              "4   CACTCTTGCGAGCATACTACTCAGGCGGAGTACTTAACGCGTTAGC...   \n",
              "5   CACACTTGCGTGCGTACTCCCCAGGCGGAACACTTAACGCGTTGGC...   \n",
              "6   TAATCTTGCGACCGTACTCCCCAGGCGGAATGCTTAATCCGTTAGG...   \n",
              "7   TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...   \n",
              "8   TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...   \n",
              "9   TAGTCTTGCGACCGTAGTCCCCAGGCGGAGTGCTTAACGCGTTAGC...   \n",
              "10  CACACTTGCGTGCGTACTCCCCAGGCGGAACACTTAACGCGTTGGC...   \n",
              "11  TAGTCTTGCGACCGTACTCCCCAGGCGGAGAACTTAACGCGTTAGC...   \n",
              "12  TAATCTTGCGACCGTACTCCCCAGGCGGTCTACTTAACGCGTTAGC...   \n",
              "13  CAGCCTTGCGACCATACTCCCCAGGCGGAACACTTAACGCTTTCGC...   \n",
              "14  TAATCTTGCGACCGTACTCCCCAGGCGGTTCATTTAATGCGTTAGC...   \n",
              "15  CATTCTTGCGAACGTACTCCCCAGGTGGGATACTTATCACTTTCGC...   \n",
              "16  TAACCTTGCGGCCGTACTCCCCAGGCGGTCAACTTATCGCGTTAGC...   \n",
              "17  TAGCCTTGCGGCCGTACTCCCCAGGCGGTACACTTATCGCGTTAGC...   \n",
              "18  TAATCTTGCGACCGTACTCCCCAGGCGGAGTGCTTAATGCGTTAGC...   \n",
              "19  CAACCTTGCGGCCGTACTCCCCAGGCGGAGTGCTTAATGCGTTAGC...   \n",
              "20  TAGCCTTGCGGCCGTACTCCCCAGGCGGATAACTTAACGCGTTAGC...   \n",
              "21  TAGCCTTGCGGCCGTACTCCCCAGGCGGACAACTTAAAACGTTAGC...   \n",
              "22  CATTCTTGCGAACGTACTCCCCAGGTGGGATACTTATCACTTTCGC...   \n",
              "23  TAATCTTGCGACCGTACTCCCCAGGCGGTCTACTTATCGCGTTAGC...   \n",
              "24  TAACCTTGCGGCCGTACTCCCCAGGCGGTGTGCTTAATGCGTTAGC...   \n",
              "25  TAGCCTTGCGGCCGTACTCCCCAGGCGGAGAACTTATCGCGTTAGC...   \n",
              "\n",
              "                          Feature ID  \n",
              "0   8331adf13c184313c464be9c7b21097c  \n",
              "1   b37b36ef2cd58aa19929a8ae4f8f4c21  \n",
              "2   b5f4256d637420120c2e71bc85666bfd  \n",
              "3   00aabc54895675bbc65b0c1fdaf1fd07  \n",
              "4   3c5ee7863e02dd5604419d803a014014  \n",
              "5   fe008f8f1f5a1776177e14bae97e10e2  \n",
              "6   3a4c4c7d2b8e5a489eb2d1fddc95e025  \n",
              "7   0c2db32fc52f027a04e433a93c0669df  \n",
              "8   5f460fa8105db752feb6166a52b135d2  \n",
              "9   6d4121b0c9e7eda284b4ab5bf571dc86  \n",
              "10  6e0740a41b665e09ed40342b621fb3b0  \n",
              "11  c6eddca0dc7ee0b43403e873ba6c8cfc  \n",
              "12  d4c1d85ff4ca5efdc346801641b2d9a2  \n",
              "13  c00ee893109657877cc78cca52003ca9  \n",
              "14  d6f1d18cadc754defc476dd14e4f4c58  \n",
              "15  ac30e74e21dd088eabfea4d8738c6279  \n",
              "16  4d0e9bcbdf46ba60e8c807aca0550f26  \n",
              "17  3046631d2ae55b33a73e710aecb4db9e  \n",
              "18  6ae71cd37d021784c75d9c4ca7de333d  \n",
              "19  c27592c0acdd2708a5b18cf790d0d1e1  \n",
              "20  a6c50d7fef8a40824552274d0e96e3c4  \n",
              "21  f67e6cc781a78a69a56d9d441411694b  \n",
              "22  c92bce272a01f65e7105823ab7957ae8  \n",
              "23  c74f9ca1eae5081f61f84934c8e22a79  \n",
              "24  b770c9ce6a8ca4de3d7085854747551c  \n",
              "25  f1a3e9623a42e1ee02b2317bb9a02a25  "
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# display(denoised_FtableDF)\n",
        "freq_seq = denoised_FtableDF.merge(denoised_SeqsDF, on='Feature ID')\n",
        "freq_seq['Feature ID'] = freq_seq.index\n",
        "freq_seq = freq_seq.reset_index(drop=True)\n",
        "display(freq_seq)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "<Figure size 640x480 with 2 Axes>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "sns.heatmap(freq_seq.iloc[:, :-2])\n",
        "plt.show()"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### Taxonomic analyses"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Import combined databases \n",
        "ids = Artifact.load('../data/Ids.qza')\n",
        "seqs = Artifact.load('../data/all_fasta.qza')"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### Alignment-based classification"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "from qiime2.plugins.feature_classifier.pipelines import classify_consensus_blast\n",
        "\n"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### Machine learning"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Data sources\n",
        "- https://www.ebi.ac.uk/ena/browser/view/PRJEB8518\n",
        "- https://docs.qiime2.org/2023.5/data-resources/  High-quality reference OTUs\n",
        "- The `see` and `v2frame` functions were obtained from [this](https://forum.qiime2.org/t/how-to-capture-a-value-from-a-summary-and-pipe-it/19783/5) link via user thermokarst"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8.15"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 4
}
