### Jellyfish microbiome analysis  


```python
from qiime2 import Metadata
import pandas as pd
from qiime2 import Artifact, sdk

pm = sdk.PluginManager()
def see(artifact):
    from_format = artifact.format
    if issubclass(from_format, sdk.plugin_manager.SingleFileDirectoryFormatBase):
        from_format = artifact.format.file.format
    return set(pm.transformers[from_format].keys())
import os
import pandas
import qiime2
import tempfile

def v2frame(viz_fp: str) -> list:
    '''viz_fp is a path to the qiime2 visualization object'''
    viz = qiime2.Visualization.load(viz_fp)
    with tempfile.TemporaryDirectory() as tmpdir:
        viz.export_data(tmpdir)
        fp = os.path.join(tmpdir, 'quality-plot.html')
        ov = os.path.join(tmpdir, 'overview.html')
        dfs = pandas.read_html(fp, index_col=0)
        df2s = pandas.read_html(ov, index_col=0)
    return dfs + df2s
```
#### Importing the data
The fastq files contain single-end reads, sequenced with the 454 GS FLX+ via a pyrosequencing approach

```bash
qiime tools import \
    --type EMPSingleEndSequences \
    --input-path jelly_data \
    --output-path jelly_seqs.qza
```


## Data sources
https://www.ebi.ac.uk/ena/browser/view/PRJEB8518
- The `see` and `v2frame` functions were obtained from [this](https://forum.qiime2.org/t/how-to-capture-a-value-from-a-summary-and-pipe-it/19783/5) link via user thermokarst
